---
title: "Early stage copy number method validation"
author: "Philip Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: TRUE
    toc_float: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries

```{r load_libraries}
# Load libraries
suppressPackageStartupMessages(suppressWarnings(library(tidyverse)))
suppressPackageStartupMessages(suppressWarnings(library(kableExtra)))
suppressPackageStartupMessages(suppressWarnings(library(biomaRt)))
suppressPackageStartupMessages(suppressWarnings(library(Biobase)))
suppressPackageStartupMessages(suppressWarnings(library(QDNAseqmod)))
suppressPackageStartupMessages(suppressWarnings(library(cowplot)))
```

### Assign int. functions

Load functions used in the focal analysis document and additional functions used in the manipulation of the RDS QDNAseq object, initially written by Geoff

```{r functions}
source("../focal_analysis/data/focal_cna_analysis_funcs.R")
source("../focal_analysis/data/nbclustfixed.R")

getSampNames<-function(abs_profiles)
{
    if(class(abs_profiles)=="QDNAseqCopyNumbers")
    {
        samps<-colnames(abs_profiles)
    }
    else
    {
        samps<-names(abs_profiles)
    }
    samps
}

getPloidy<-function(abs_profiles)
{
  out<-c()
  samps<-getSampNames(abs_profiles)
  for(i in samps)
  {
    if(class(abs_profiles)=="QDNAseqCopyNumbers")
    {
      segTab<-getSegTable(abs_profiles[,which(colnames(abs_profiles)==i)])
    }
    else
    {
      segTab<-abs_profiles[[i]]
      colnames(segTab)[4]<-"segVal"
    }
    segLen<-(as.numeric(segTab$end)-as.numeric(segTab$start))
    ploidy<-sum((segLen/sum(segLen))*as.numeric(segTab$segVal))
    out<-c(out,ploidy)
  }
  data.frame(out,stringsAsFactors = F)
}

```

Load the focal gene table data which describes the genes of interest and the meta data for the BriTROC study copy number profiles

```{r read_meta}
focal_genes_table <- read.table("../focal_analysis/data/focal_CNA_pseudo_bed.txt",header = T,sep = "\t")

meta.data <- read.table("../../copy_number_signatures/britroc_30kb_signature_data_meta.tsv",header = TRUE,sep = "\t")
# patient_data <- read.table("../../britroc_cohort_patient_data.tsv",header = T,sep = "\t")
```

Load list of protein coding genes used for bin extraction and annotation in focal copy number analysis

```{r biomart}
ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl",GRCh = 37)
protein_coding_genes <- getBM(attributes = c("chromosome_name","start_position","end_position","hgnc_symbol"),
                                             filters = "biotype",
                                             values = "protein_coding",
                                             mart = ensembl,
                                             useCache = FALSE)
protein_coding_genes <- protein_coding_genes[protein_coding_genes$chromosome_name %in% seq.int(1,22,1),]
protein_coding_genes <- protein_coding_genes[protein_coding_genes$hgnc_symbol != "",]
```

Load absolute copy number data related to BriTROC paper 2

```{r read_abs}
abs_data <- readRDS("../../absolute_POST_down_sampling/britroc_30kb_ds_absCopyNumber.rds")
abs_data <- abs_data[,which(colnames(abs_data) %in% meta.data$SAMPLE_ID[meta.data$use == "TRUE"])]
```

Get ploidy for britroc2 samples

```{r phil_set_ploidy}
phil_ploidys <- pData(abs_data)$ploidy
names(phil_ploidys) <- pData(abs_data)$name
```

Load absolute copy number data related the late stage patients in the early to late paper

```{r zhao_read_abs}
zhao_abs_data <- readRDS("data/30kb_ds_absCopyNumber_late_stage.rds")
```

Get ploidy for early samples

```{r zhao_set_ploidy}
zhao_ploidys <- pData(zhao_abs_data)$ploidy
names(zhao_ploidys) <- pData(zhao_abs_data)$name
```

Extract overlapping samples from these sets and compare ploidy

```{r compare_ploidys}
inter_ploidy <- intersect(names(zhao_ploidys),names(phil_ploidys))
all(zhao_ploidys[names(zhao_ploidys) %in% inter_ploidy] == phil_ploidys[names(phil_ploidys) %in% inter_ploidy])
length(inter_ploidy)
```

Check if ploidy of matched samples are correlated

```{r compare_ploidies_plot}
ploidy_plot <- data.frame(early=c(zhao_ploidys[names(zhao_ploidys) %in% inter_ploidy]),
                          britroc=c(phil_ploidys[names(phil_ploidys) %in% inter_ploidy]))
ggplot(ploidy_plot) +
  geom_point(aes(early,britroc)) +
  theme_bw()
```

Extract pre-computed copy number counts for all genes in the BriTROC paper 2 data

```{r extract_phil}
phil_extracted <- readRDS("../focal_analysis/data/extracted_output.RDS")
```

## BriTROC paper 2 method
Call CNA using the functions in the focal analysis

```{r cna_class_by_ploidy_gene_phil}
ploidys <- phil_ploidys
phil_cna_calls <- get_cna_calls(data = phil_extracted)
phil_cna_calls_sub <- phil_cna_calls[,colnames(phil_cna_calls) %in% inter_ploidy]
ploidys <- NULL
```

Focal rates in the overlapping sample subset in the BriTROC paper 2 data

```{r count_focal_amp_phil}
phil_focal_cna_rates <- get_cna_rates(data = phil_cna_calls_sub,genes = focal_genes_table$Gene.name)
phil_focal_cna_rates[order(phil_focal_cna_rates$rate,decreasing = T),]
```

Perform the same copy number extraction and calling on the overlapping samples in the early paper using BriTROC methodology

```{r extract_zhao}
if(file.exists("data/extracted_output.RDS")){
  zhao_extracted <- readRDS("data/extracted_output.RDS")
} else {
  zhao_extracted <- extract_abs_segs(abs_data = zhao_abs_data,genes = protein_coding_genes) # long execution times
  saveRDS(zhao_extracted,file = "data/extracted_output.RDS") 
}
```

```{r cna_class_by_ploidy_gene_zhao}
ploidys <- zhao_ploidys
zhao_cna_calls <- get_cna_calls(data = zhao_extracted)
zhao_cna_calls_sub <- zhao_cna_calls[,colnames(zhao_cna_calls) %in% inter_ploidy]
ploidys <- NULL
```

```{r count_focal_amp_zhao}
zhao_focal_cna_rates <- get_cna_rates(data = zhao_cna_calls_sub,genes = focal_genes_table$Gene.name)
zhao_focal_cna_rates[order(zhao_focal_cna_rates$rate,decreasing = T),]
```

Plot gene alteration rates in both sets of overlapping samples using BriTROC paper 2 methodology

```{r compare_focal_rates}
phil_focal_cna_rates$data <- rep("britroc",times=nrow(phil_focal_cna_rates))
zhao_focal_cna_rates$data <- rep("early",times=nrow(zhao_focal_cna_rates))

combined_focal_rates <- rbind(phil_focal_cna_rates,zhao_focal_cna_rates)
ggplot(combined_focal_rates) +
  geom_col(aes(gene,rate,fill=data),position = "dodge") +
  facet_wrap(. ~ group) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5))
```

Alternative implementation of the gene alteration rate calling for both overlapping sets.

```{r calls_analysis}
zhao_cna_call_sub <- zhao_cna_calls[rownames(zhao_cna_calls) %in% focal_genes_table$Gene.name,colnames(zhao_cna_calls) %in% inter_ploidy]
zhao_recalled_cna_calls <- do.call(rbind,apply(zhao_cna_call_sub,MARGIN = 1,FUN = function(x) data.frame(N=sum(x ==  "N"),
                                                                                               AMP=sum(x == "AMP"),
                                                                                               DEL=sum(x == "DEL"),stringsAsFactors = F))) %>%
                      rownames_to_column("gene") %>%
                      mutate(amp_rate = (AMP / (N+AMP+DEL)) * 100) %>%
                      mutate(del_rate = (DEL / (N+AMP+DEL)) * 100) %>%
                      arrange(gene)

phil_recalled_cna_calls <- do.call(rbind,apply(phil_cna_calls_sub[rownames(phil_cna_calls_sub) %in% focal_genes_table$Gene.name,],MARGIN = 1,FUN = function(x) data.frame(N=sum(x ==  "N"),
                                                                                               AMP=sum(x == "AMP"),
                                                                                               DEL=sum(x == "DEL"),stringsAsFactors = F))) %>%
                      rownames_to_column("gene") %>%
                      mutate(amp_rate = (AMP / (N+AMP+DEL)) * 100) %>%
                      mutate(del_rate = (DEL / (N+AMP+DEL)) * 100) %>%
                      arrange(gene)

identical(zhao_recalled_cna_calls,phil_recalled_cna_calls)
```

## Compare method outcomes
Load the pre-computed copy number counts by gene in the early paper

```{r load_zhao_method}
zhao_high_ploidy <- read.table("data/high_ploidy_britroc_abscn.txt",header = T,sep = "\t")
zhao_low_ploidy <- read.table("data/low_ploidy_britroc_abscn.txt",header = T,sep = "\t")

zhao_data <- cbind(zhao_high_ploidy,zhao_low_ploidy)
zhao_data$group <- rep("early_method",times=nrow(zhao_high_ploidy))

zhao_data <- zhao_data[,colnames(zhao_data) %in% c("name","group",inter_ploidy)]
zhao_data <- zhao_data[,colnames(zhao_data) %in% c("name","group",inter_ploidy)]
head(zhao_data)
```

Subset copy number data in both sets to overlapping samples

```{r compare_cn_reformat}
phil_extracted_sub <- phil_extracted[rownames(phil_extracted) %in% zhao_high_ploidy$name,colnames(phil_extracted) %in% inter_ploidy]
phil_extracted_sub <- phil_extracted_sub %>%
  rownames_to_column("name") %>%
  mutate(group = rep("britroc_data",times=nrow(.))) %>%
  arrange(name)

zhao_extracted_sub <- zhao_extracted[rownames(zhao_extracted) %in% zhao_high_ploidy$name,colnames(zhao_extracted) %in% inter_ploidy]
zhao_extracted_sub <- zhao_extracted_sub %>%
  rownames_to_column("name") %>%
  mutate(group = rep("britroc_method",times=nrow(.))) %>%
  arrange(name)

extracted_reformat <- rbind(phil_extracted_sub,zhao_extracted_sub,zhao_data) %>%
  relocate(name,group) %>%
  pivot_longer(names_to = "sample",cols = 3:ncol(.)) %>%
  arrange(sample,name) %>%
  group_by(sample,name) %>%
  mutate(diff_pd_zm = signif(value[group == "britroc_data"] - value[group == "early_method"],digits = 3)) %>%
  mutate(diff_pd_pm = signif(value[group == "britroc_data"] - value[group == "britroc_method"],digits = 3))

extracted_reformat$sample <- factor(extracted_reformat$sample,
                             levels = unique(extracted_reformat$sample[order(extracted_reformat$diff_pd_zm,decreasing = T)]))
head(extracted_reformat)
```

```{r compare_cn_plot_1}
ggplot(extracted_reformat) +
  geom_line(aes(group,value,group = sample),alpha = 0.5) +
  geom_point(aes(group,value,group = sample,color=group)) +
  facet_wrap(. ~ name,
             scale="free_y") +
  theme_bw()
```

Copy number value deviations between britroc_method and early_method for number of genes exist

```{r compare_cn_plot_2}
extracted_reformat_sub <- extracted_reformat[extracted_reformat$group == "britroc_data",] %>%
                            dplyr::select(name,sample,diff_pd_zm,diff_pd_pm) %>%
                            pivot_longer(cols = c("diff_pd_zm","diff_pd_pm"),names_to = "difference",values_to = "value")

ggplot(extracted_reformat_sub) +
  #geom_boxplot(aes(name,value,fill=difference)) +
  geom_point(aes(name,value,color=difference),position = position_jitterdodge(jitter.width = 0.2)) +
  geom_hline(yintercept = 0) +
  facet_wrap(. ~ name,scales = "free_x") +
  theme_bw() +
  theme(axis.text.x = element_blank())
```

## Examine early method

```{r zhao_bin_to_gene}
late<-readRDS("data/30kb_ds_absCopyNumber_late_stage.rds")
digits<-3 # added by PS
# And using the same method for late stage cohort
locations <- fData(late) %>%
  rownames_to_column(var = "id") %>%
  as_tibble() %>%
  dplyr::select(id, chromosome, start, end) %>%
  mutate_at(vars(start, end), as.integer) %>%
  mutate(chromosome = factor(chromosome, levels = unique(chromosome)))
#locations

samples <- sort(sampleNames(late))
number_of_samples <- length(samples)
#number_of_samples

for (sample_index in 1:number_of_samples) {
  sample <- samples[sample_index]
  cat(sample_index, "/", number_of_samples, " ", sample, "\n", sep = "")
  
  copy_number_for_sample <- late[, sample]
  
  copy_number_values <- assayDataElement(copy_number_for_sample, "copynumber")
  if (!all(rownames(copy_number_values) == locations$id)) stop("Error extracting copy number")
  
  segmented_values <- assayDataElement(copy_number_for_sample, "segmented")
  if (!all(rownames(segmented_values) == locations$id)) stop("Error extracting segmented copy number")
  
  copy_number_for_sample <- locations %>%
    mutate(sample = sample, copy_number = copy_number_values[,1], segmented = segmented_values[,1])
  
  copy_number_for_sample <- copy_number_for_sample %>%
    mutate_at(vars(copy_number, segmented), round,digits) %>%
    mutate(copy_number = ifelse(copy_number == 0, 0, copy_number)) %>%
    mutate(segmented = ifelse(segmented == 0, 0, segmented)) %>%
    dplyr::select(sample, chromosome, start, end, copy_number, segmented)
}

gene<-read.table("data/genes.txt",header=T,sep="\t")
head(gene)
gene<-as.data.frame(gene)

test_z <- data.frame() # Added PS
for (sample_index in 1:number_of_samples) {
  sample <- samples[sample_index]
  cat(sample_index, "/", number_of_samples, " ", sample, "\n", sep = "")
  
  copy_number_for_sample <- late[, sample]
  
  copy_number_values <- assayDataElement(copy_number_for_sample, "copynumber")
  if (!all(rownames(copy_number_values) == locations$id)) stop("Error extracting copy number")
  
  segmented_values <- assayDataElement(copy_number_for_sample, "segmented")
  if (!all(rownames(segmented_values) == locations$id)) stop("Error extracting segmented copy number")
  
  copy_number_for_sample <- locations %>%
    mutate(sample = sample, copy_number = copy_number_values[,1], segmented = segmented_values[,1])
  
  copy_number_for_sample <- copy_number_for_sample %>%
    mutate_at(vars(copy_number, segmented), round,digits) %>%
    mutate(copy_number = ifelse(copy_number == 0, 0, copy_number)) %>%
    mutate(segmented = ifelse(segmented == 0, 0, segmented)) %>%
    dplyr::select(sample, chromosome, start, end, copy_number, segmented)
  
  new_gene<-merge(copy_number_for_sample,gene,by="chromosome",all=F)
  
  x<-new_gene[which((new_gene$start.y-new_gene$start.x)>=0&(new_gene$start.y-new_gene$end.x)<=0&(new_gene$end.y-new_gene$end.x)<=0&(new_gene$end.y-new_gene$start.x)>=0),]
  y<-new_gene[which((new_gene$start.y-new_gene$start.x)/10^8*(new_gene$start.y-new_gene$end.x)/10^8*(new_gene$end.y-new_gene$start.x)/10^8*(new_gene$end.y-new_gene$end.x)/10^8<0),]
  z<-bind_rows(x,y)
  
  original_z<-z %>%
    as_tibble() %>%
    dplyr::select(sample,name,segmented)
  
  #test_z<-cbind(test_z,original_z) 
  test_z<-rbind(test_z,original_z) ## ADDED PS
}

## ADDED PS to replicate undocumented format alteration (uses mean)
test_z <- test_z %>%
  distinct() %>%
  group_by(sample,name) %>%
  summarise(across(segmented,mean)) %>%
  pivot_wider(id_cols = sample,names_from = name,values_from = segmented) %>%
  column_to_rownames(var = "sample")

test_z <- t(test_z)
test_z <- as.data.frame(test_z) %>%
            mutate(name = rownames(.)) %>%
            relocate(name)
rownames(test_z) <- test_z$name
head(test_z)
#write.table(test_z,"late_sample_segemented_copy_number.txt",sep = "\t")
```

```{r zhao_calling_code}
ploidy<-read.table("data/high_ploidy_britroc_cn.txt",sep = "\t",header = T)
ploidy<-as.data.frame(ploidy)
ploidy<-ploidy[,-1]

late1<-read.table("data/low_ploidy_britroc_abscn.txt",sep="\t",header = T)
late2<-read.table("data/high_ploidy_britroc_abscn.txt",sep = "\t",header = T)
late1<-as.data.frame(late1)
rownames(late1)<-late1[,1]
final_late1<-late1[,-1]
late2<-as.data.frame(late2)
rownames(late2)<-late2[,1]
final_late2<-late2[,-1]

# For our definition of gene amplification and deletions.
# When the ploidy<=2.7,choose the ab copy number >=5 as amplification
# When average genome ploidy <= 2.7 AND total copy number = 0
final_late1<-round(final_late1[,1:26],digits=0)
a1<-replace(final_late1,final_late1>=5,"Amp")
b1<-replace(a1,a1<=0,"Del")
c1<-replace(b1,b1>=1&b1<=4,"")

# When the ploidy>2.7,choose the ab copy number >=9 as amplification
# When average genome ploidy >2.7 AND total copy number<(average genome ploidy-2.7)
final_late2 <-round(final_late2[,1:42],digits=0)
a2<-replace(final_late2 ,final_late2 >=9,"Amp")
b2<-(ploidy-2.7)
c2<-replace(a2,a2<b2,"Del")
d2<-replace(c2,c2>=b2&c2<9,"")
e4<-cbind(d2,c1)

#head(e4)
e4[is.na(e4)] = ""
#head(e4)
e4= as.matrix(e4)
#e4

d2_reformat <- do.call(rbind,apply(d2,MARGIN = 1,FUN = function(x) data.frame(N=sum(x ==  ""),
                                                                                               AMP=sum(x == "Amp"),
                                                                                               DEL=sum(x == "Del")))) %>%
                      rownames_to_column("gene") %>%
                      mutate(amp_rate = (AMP / (N+AMP+DEL)) * 100) %>%
                      mutate(del_rate = (DEL / (N+AMP+DEL)) * 100) %>%
                      arrange(gene)
d2_reformat
zhao_recalled_cna_calls
```

## calling method (Min Repro)

Error was identified in the calling of copy number alterations.

```{r minimal_repo}
sample_ploidys <- c(sample1=2.4,sample2=1.5,sample3=3.6,sample4=3.2)
minimal_input <- data.frame(sample1 = c(6.2,3.1,0,2.5),
           sample2 = c(2.1,0,2.8,2.3),
           sample3 = c(5.2,9.2,2.4,0.8),
           sample4 = c(2.3,12.1,2.9,3.3),
           row.names = paste0("gene",LETTERS[1:4]))
minimal_input
```

```{r expected_out}
sample_ploidys
minimal_input

minimal_out <- data.frame(sample1 = c("AMP","N","DEL","N"),
           sample2 = c("N","DEL","N","N"),
           sample3 = c("N","AMP","N","DEL"),
           sample4 = c("N","AMP","N","N"),
           row.names = paste0("gene",LETTERS[1:4]))
minimal_out
```

```{r test_minimal_ps}
get_cna_calls <- function(data=NULL){
  # COSMIC definitions
  #Gain:
  ## average genome ploidy <= 2.7 AND total copy number >= 5
  ## OR average genome ploidy > 2.7 AND total copy number >= 9
  #Loss:
  ## average genome ploidy <= 2.7 AND total copy number = 0
  ## OR average genome ploidy > 2.7 AND total copy number < ( average genome ploidy - 2.7 )
  cna_calls <- do.call(cbind,lapply(colnames(data),FUN = function(x){
    ploidy <- sample_ploidys[x]
    abs_col <- data[,x]
    if(ploidy > 2.7){
      abs_col <- ifelse(abs_col >= 9,"AMP",ifelse(abs_col < (ploidy - 2.7),"DEL","N"))
    } else {
      abs_col <- ifelse(abs_col >= 5,"AMP",ifelse(abs_col <= 0,"DEL","N"))
    }
    abs_col <- data.frame(sample=abs_col)
    colnames(abs_col) <- x
    return(abs_col)
  }))
  rownames(cna_calls) <- rownames(data)
  return(cna_calls)
}

min_test_ps <- get_cna_calls(data = minimal_input)
min_test_ps
```

```{r test_minimal_ps_check}
identical(minimal_out,min_test_ps)
```

```{r zhao_code_test}
ploidy_z<-data.frame(rbind(sample_ploidys,sample_ploidys,sample_ploidys,sample_ploidys))
late2<-minimal_input
#late2<-as.data.frame(late2)
#rownames(late2)<-late2[,1]
final_late2<-minimal_input

# When the ploidy>2.7,choose the ab copy number >=9 as amplification
# When average genome ploidy >2.7 AND total copy number<(average genome ploidy-2.7)
final_late2 <-round(final_late2,digits=0)
a2<-replace(final_late2 ,final_late2 >=9,"Amp")
b2<-(ploidy_z-2.7)
c2<-replace(a2,a2<b2,"Del")
d2<-replace(c2,c2>=b2&c2<9,"")

# alter empty to be N
d2[d2 == ""] <- "N"
min_test_z <- d2
min_test_z
```

```{r test_minimal_zc_check}
identical(minimal_out,min_test_z)
```

### Gene bin selection

```{r bin_selection_ps}
extract_abs_segs <- function(abs_data = x,genes = NULL){
  sample_n <- length(colnames(abs_data))
  to_use <- fData(abs_data)$use
  all_segs <- data.frame()
  gene_bins <- get_gene_bins(abs_data = abs_data,genes = genes)
  sample_segs <- c()
  for(i in 1:sample_n){
    cn_obj <- abs_data[to_use,i]
    #ploidy <- pData(cn_obj)$ploidy
    sample_name <- colnames(cn_obj)
    segments <- assayDataElement(cn_obj,"segmented")
    gene_segs <- c()
    for(j in 1:nrow(gene_bins)){
      bins <- gene_bins[j,]
      start <- bins$start_idx
      end <- bins$end_idx
      seg_mean <- mean(segments[c(start:end)])
      gene_segs <- cbind(gene_segs,seg_mean)
    }
    gene_segs <- as.data.frame(gene_segs)
    colnames(gene_segs) <- gene_bins$gene
    sample_segs <- rbind(sample_segs,gene_segs)
  }
  rownames(sample_segs) <- colnames(abs_data)
  return(as.data.frame(t(sample_segs)))
}

get_gene_bins <- function(abs_data = x,genes = y){
  to_use <- fData(abs_data)$use
  cn_obj <- abs_data[to_use,]
  bin_pos <- fData(cn_obj)[,c("chromosome","start","end")]
  gene_bins <- c()
  for(i in 1:nrow(genes)){
    chr <- as.character(genes[i,1])
    start <- as.numeric(genes[i,2])
    end <- as.numeric(genes[i,3])
    g <- as.character(genes[i,4])
    gene_chr_pos <- bin_pos[bin_pos$chromosome == chr,]
    ## Remove genes outside of binned regions
    if(start < gene_chr_pos$start[1] & end < gene_chr_pos$start[1]){
      cat(paste0(g," outside binned genome... skipping\n"))
      next()
    }
    if(start > gene_chr_pos$end[nrow(gene_chr_pos)] & end > gene_chr_pos$end[nrow(gene_chr_pos)]){
      cat(paste0(g," outside binned genome... skipping\n"))
      next()
    }
    min_start <- min(which(min(abs(gene_chr_pos$start - start)) == abs(gene_chr_pos$start - start)))
    min_end <- max(which(min(abs(gene_chr_pos$end - end)) == abs(gene_chr_pos$end - end)))
    # min and max are used for min_start and min_end due to a few (< 5) genes having end and start points equi-distance from bins
    # Additionally, genes spanning the first or last available bin are restricted to only a complete bin as no information is available
    # outside of those bins for cn state
    if(gene_chr_pos$start[min_start] > start & min_start != 1){
      min_start <- min_start - 1
    }
    if(gene_chr_pos$end[min_end] < end & min_end != length(gene_chr_pos$end)){
      min_end <- min_end + 1
    }
    
    index_min <- which(bin_pos$chromosome == chr & bin_pos$start == gene_chr_pos$start[min_start])
    index_max <- which(bin_pos$chromosome == chr & bin_pos$end == gene_chr_pos$end[min_end])
    start.dist <- start - gene_chr_pos$start[min_start]
    end.dist <- gene_chr_pos$end[min_end] - end
    ## Removes rows where either the gene start or end fall outside available bins
    if(end.dist < 0 | start.dist < 0){
      cat(paste0(g," single end out of range... skipping\n"))
      next()
    }
    gene_pos <- data.frame(start_idx=index_min,
                           end_idx=index_max,
                           gene=g,
                           start.dist=start.dist,
                           end.dist=end.dist)
    gene_bins <- rbind(gene_bins,gene_pos)
  }
  return(gene_bins)
}

CCCNE1_ps <- extract_abs_segs(abs_data = abs_data[,colnames(abs_data) %in% inter_ploidy],genes = focal_genes_table[focal_genes_table$Gene.name == "CCNE1",])
CCCNE1_ps
```

```{r bin_selection_zc}
digits<-3 # added by PS

late <- late[,colnames(late) %in% inter_ploidy]
samples <- sort(sampleNames(late))
number_of_samples <- length(samples)
gene <- gene[gene$name == "CCNE1",]

test_z <- data.frame() # Added PS
for (sample_index in 1:number_of_samples) {
  sample <- samples[sample_index]
  cat(sample_index, "/", number_of_samples, " ", sample, "\n", sep = "")
  
  copy_number_for_sample <- late[, sample]
  
  copy_number_values <- assayDataElement(copy_number_for_sample, "copynumber")
  if (!all(rownames(copy_number_values) == locations$id)) stop("Error extracting copy number")
  
  segmented_values <- assayDataElement(copy_number_for_sample, "segmented")
  if (!all(rownames(segmented_values) == locations$id)) stop("Error extracting segmented copy number")
  
  copy_number_for_sample <- locations %>%
    mutate(sample = sample, copy_number = copy_number_values[,1], segmented = segmented_values[,1])
  
  copy_number_for_sample <- copy_number_for_sample %>%
    mutate_at(vars(copy_number, segmented), round,digits) %>%
    mutate(copy_number = ifelse(copy_number == 0, 0, copy_number)) %>%
    mutate(segmented = ifelse(segmented == 0, 0, segmented)) %>%
    dplyr::select(sample, chromosome, start, end, copy_number, segmented)
  
  new_gene<-merge(copy_number_for_sample,gene,by="chromosome",all=F)
  
  x<-new_gene[which((new_gene$start.y-new_gene$start.x)>=0&(new_gene$start.y-new_gene$end.x)<=0&(new_gene$end.y-new_gene$end.x)<=0&(new_gene$end.y-new_gene$start.x)>=0),]
  y<-new_gene[which((new_gene$start.y-new_gene$start.x)/10^8*(new_gene$start.y-new_gene$end.x)/10^8*(new_gene$end.y-new_gene$start.x)/10^8*(new_gene$end.y-new_gene$end.x)/10^8<0),]
  z<-bind_rows(x,y)
  
  original_z<-z %>%
    as_tibble() %>%
    dplyr::select(sample,name,segmented)
  
  #test_z<-cbind(test_z,original_z) 
  test_z<-rbind(test_z,original_z) ## ADDED PS
}

## ADDED PS to replicate undocumented format alteration (uses mean)
test_z <- test_z %>%
  distinct() %>%
  group_by(sample,name) %>%
  summarise(across(segmented,mean)) %>%
  pivot_wider(id_cols = sample,names_from = name,values_from = segmented) %>%
  column_to_rownames(var = "sample")

test_z <- t(test_z)
test_z <- as.data.frame(test_z) %>%
            mutate(name = rownames(.)) %>%
            relocate(name)
rownames(test_z) <- test_z$name
test_z <- test_z[,-1]
test_z
```

```{r test_bins}
CCNE1_bins_ps <- get_gene_bins(abs_data = abs_data[,colnames(abs_data) %in% inter_ploidy],genes = focal_genes_table[focal_genes_table$Gene.name == "CCNE1",])[,c(1:2)]
CCNE1_bins_zc <- get_gene_bins(abs_data = abs_data[,colnames(abs_data) %in% inter_ploidy],genes = gene[gene$name == "CCNE1",c(2:4,1)])[,c(1:2)]

fData(abs_data)[c(CCNE1_bins_ps$start_idx:CCNE1_bins_ps$end_idx),]
fData(abs_data)[c(CCNE1_bins_zc$start_idx:CCNE1_bins_zc$end_idx),]

fData(late)[c(CCNE1_bins_ps$start_idx:CCNE1_bins_ps$end_idx),]
fData(late)[c(CCNE1_bins_zc$start_idx:CCNE1_bins_zc$end_idx),]

ps_bins_psCN <- assayDataElement(abs_data[,colnames(abs_data) %in% inter_ploidy],elt = "segmented")[c(CCNE1_bins_ps$start_idx:CCNE1_bins_ps$end_idx),]
ps_bins_zcCN <- assayDataElement(late,elt = "segmented")[c(CCNE1_bins_ps$start_idx:CCNE1_bins_ps$end_idx),]

ps_bins <- as.data.frame(cbind(ps_bins_psCN,ps_bins_zcCN)) %>%
  mutate(diff = ps_bins_psCN - ps_bins_zcCN)

zc_bins_psCN <- colMeans(assayDataElement(abs_data[,colnames(abs_data) %in% inter_ploidy],elt = "segmented")[c(CCNE1_bins_zc$start_idx:CCNE1_bins_zc$end_idx),])

zc_bins_zcCN <- colMeans(assayDataElement(late,elt = "segmented")[c(CCNE1_bins_zc$start_idx:CCNE1_bins_zc$end_idx),])

zc_bins <- as.data.frame(cbind(zc_bins_psCN,zc_bins_zcCN)) %>%
  mutate(diff = zc_bins_psCN - zc_bins_zcCN)

ps_bins
zc_bins
```

Differences in gene copy number values due to incorrectly using hg38 genome locations. This explains the variability in the differences, and the semi-random effect seen across samples, where some genes are close in different builds whilst others are not.

```{r CCNE1_loc}
CCNE1_hg19 <- c(chromosome=19,start=30302898,end=30315219,name="CCNE1")
CCNE1_hg38 <- c(chromosome=19,start=29811991,end=29824312,name="CCNE1")

ps_genes <- focal_genes_table[focal_genes_table$Gene.name == "CCNE1",c(1:4)]
colnames(ps_genes) <- c("chromosome","start","end","name")
zc_genes <- gene[gene$name == "CCNE1",c(2:4,1)]
CCNE1_comb <- rbind(CCNE1_hg19,CCNE1_hg38,ps_genes,zc_genes)
rownames(CCNE1_comb) <- c("hg19","hg38","ps","zc")
CCNE1_comb
```

```{r original_zc_implementation}
## original 
ploidy_early<-read.table("data/early_ploidy_of_samples.txt",sep = "\t",header = T)
ploidy_early<-as.data.frame(ploidy_early)
ploidy_h1<-ploidy_early[which((ploidy_early$ploidy-2.7)>0),]
rownames(ploidy_h1)<-ploidy_h1[,1]
ploidy_l1<-ploidy_early[which((ploidy_early$ploidy-2.7)<=0),]
rownames(ploidy_l1)<-ploidy_l1[,1]

final_early<-read.table("data/early_sample_segemeted_copy_number.txt",sep = "\t",header = T)
rownames(final_early)<-final_early[,1]
final_early<-final_early[,-1]
a<-round(final_early[,1:25],digits=0)

h<-a[,colnames(a)%in%ploidy_h1$SAMPLE_ID]
l<-a[,colnames(a)%in%ploidy_l1$SAMPLE_ID]

# Using ploidy to test the Amplification and deletion of each sample.
a1<-replace(l,l>=5,"Amp")
b1<-replace(a1,a1<=0,"Del")
c1<-replace(b1,b1>0&b1<5,"")
a2<-replace(h,h>0.6&h<=9,"")
d<-cbind(c1,a2)

head(d)
d[is.na(d)] = ""
head(d)
d= as.matrix(d)
d
col = c("Del" = "red", "Amp" = "blue")
alter_fun = list(
  background = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
  },
  # big blue
  Del = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), 
              gp = gpar(fill = col["Del"], col = NA))
  },
  # bug red
  Amp = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), 
              gp = gpar(fill = col["Amp"], col = NA))
  }
)
column_title1 = "early stage cohort"
heatmap_legend_param = list(title = "Alterations", at = c("Del", "Amp"), 
                            labels = c("Deletion", "Amplification"))

# The figure for the late stage cohort
# Loss:average genome ploidy <= 2.7 AND total copy number = 0;average genome ploidy > 2.7 AND total copy number < ( average genome ploidy - 2.7 )
ploidy<-read.table("data/high_ploidy_britroc_cn.txt",sep = "\t",header = T)
ploidy<-as.data.frame(ploidy)
ploidy<-ploidy[,-1]

late1<-read.table("data/low_ploidy_britroc_abscn.txt",sep="\t",header = T)
late2<-read.table("data/high_ploidy_britroc_abscn.txt",sep = "\t",header = T)
late1<-as.data.frame(late1)
rownames(late1)<-late1[,1]
final_late1<-late1[,-1]
late2<-as.data.frame(late2)
rownames(late2)<-late2[,1]
final_late2<-late2[,-1]

# For our definition of gene amplification and deletions.
# When the ploidy<=2.7,choose the ab copy number >=5 as amplification
# When average genome ploidy <= 2.7 AND total copy number = 0
final_late1<-round(final_late1[,1:26],digits=0)
a1<-replace(final_late1,final_late1>=5,"Amp")
b1<-replace(a1,a1<=0,"Del")
c1<-replace(b1,b1>=1&b1<=4,"")

# When the ploidy>2.7,choose the ab copy number >=9 as amplification
# When average genome ploidy >2.7 AND total copy number<(average genome ploidy-2.7)
final_late2 <-round(final_late2[,1:42],digits=0)
a2<-replace(final_late2 ,final_late2 >=9,"Amp")
b2<-(ploidy-2.7)
c2<-replace(a2,a2<b2,"Del")
d2<-replace(c2,c2>=b2&c2<9,"")
e4<-cbind(d2,c1)

head(e4)
e4[is.na(e4)] = ""
head(e4)
e4= as.matrix(e4)
```

```{r new_ps_implementation}
get_cna_calls <- function(data=NULL){
  # COSMIC definitions
  #Gain:
  ## average genome ploidy <= 2.7 AND total copy number >= 5
  ## OR average genome ploidy > 2.7 AND total copy number >= 9
  #Loss:
  ## average genome ploidy <= 2.7 AND total copy number = 0
  ## OR average genome ploidy > 2.7 AND total copy number < ( average genome ploidy - 2.7 )
  cna_calls <- do.call(cbind,lapply(colnames(data),FUN = function(x){
    ploidy <- ploidys[x]
    abs_col <- data[,x]
    if(ploidy > 2.7){
      abs_col <- ifelse(abs_col >= 9,"AMP",ifelse(abs_col < (ploidy - 2.7),"DEL","N"))
    } else {
      abs_col <- ifelse(abs_col >= 5,"AMP",ifelse(abs_col <= 0,"DEL","N"))
    }
    abs_col <- data.frame(sample=abs_col)
    colnames(abs_col) <- x
    return(abs_col)
  }))
  rownames(cna_calls) <- rownames(data)
  return(cna_calls)
}

early <- readRDS("data/30kb_ds_absCopyNumber_early_stage.rds")
late <- readRDS("data/30kb_ds_absCopyNumber_late_stage.rds")

e_ploidy <- as.numeric(ploidy_early$ploidy)
names(e_ploidy) <- ploidy_early$SAMPLE_ID
l_ploidy <- as.double(ploidy[1,])
names(l_ploidy) <- colnames(ploidy)

identical(dim(fData(early)),dim(fData(late)))
zc_abs_data <- combine(early,late)

ploidys <- pData(zc_abs_data)$ploidy
names(ploidys) <- pData(zc_abs_data)$name

zc_segs <- extract_abs_segs(zc_abs_data,genes = focal_genes_table)
zc_calls <- get_cna_calls(data = zc_segs)

zc_early_rates <- get_cna_rates(zc_calls[,colnames(zc_calls) %in% colnames(early)]) %>%
                    mutate(cohort = rep("early",times=nrow(.)))
zc_late_rates <- get_cna_rates(zc_calls[,colnames(zc_calls) %in% colnames(late)]) %>%
                    mutate(cohort = rep("late",times=nrow(.)))

new_combined_rates <- rbind(zc_early_rates,zc_late_rates) %>%
                        mutate(method = rep("new",times=nrow(.)))
```

```{r old_genes_new_calls}
late_g <- late2[,-1]
early_g <- final_early

combind_oldnew <- cbind(early_g,late_g)
colnames(combind_oldnew) <- gsub(x = colnames(combind_oldnew),pattern = "\\.",replacement = "-")
combind_oldnew <- get_cna_calls(data = combind_oldnew)

zc_early_rates_oldnw <- get_cna_rates(combind_oldnew[,colnames(combind_oldnew) %in% colnames(early)]) %>%
                    mutate(cohort = rep("early",times=nrow(.)))
zc_late_rates_oldnw <- get_cna_rates(combind_oldnew[,colnames(combind_oldnew) %in% colnames(late)]) %>%
                    mutate(cohort = rep("late",times=nrow(.)))

combind_earlylate_oldnew <- rbind(zc_early_rates_oldnw,zc_late_rates_oldnw) %>%
                              mutate(method = rep("old_newcalling",times=nrow(.)))
```

```{r old_method_collate}
## Added PS
early_late_comb_calling <- cbind(d,e4)
early_late_comb_calling[early_late_comb_calling == ""] <- "N"
early_late_comb_calling[early_late_comb_calling == "Amp"] <- "AMP"
early_late_comb_calling[early_late_comb_calling == "Del"] <- "DEL"

early_late_comb_calling <- as.data.frame(early_late_comb_calling)

orig_early_rates <- get_cna_rates(early_late_comb_calling[,colnames(early_late_comb_calling) %in% colnames(early)]) %>%
                    mutate(cohort = rep("early",times=nrow(.)))
orig_late_rates <- get_cna_rates(early_late_comb_calling[,colnames(early_late_comb_calling) %in% colnames(late)]) %>%
                    mutate(cohort = rep("late",times=nrow(.)))

old_combined_rates <- rbind(orig_early_rates,orig_late_rates) %>%
                        mutate(method = rep("old",times=nrow(.)))
```

```{r plot_diff}
combined_rates <- rbind(old_combined_rates,new_combined_rates,combind_earlylate_oldnew)

ggplot(combined_rates) +
  geom_col(aes(gene,rate,fill=method),position="dodge") +
  facet_grid(rows = vars(cohort),cols = vars(group)) +
  theme_bw()
```

```{r session_info}
sessionInfo()
```
