---
title: "Analysis of absolute CNAs - BriTROC 30kb - focal"
author: "Philip Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: TRUE
    toc_float: TRUE
---

## Initial steps

### knitr options

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries

```{r load_libraries}
# Load libraries
suppressPackageStartupMessages(suppressWarnings(library(tidyverse)))
suppressPackageStartupMessages(suppressWarnings(library(kableExtra)))
suppressPackageStartupMessages(suppressWarnings(library(biomaRt)))
suppressPackageStartupMessages(suppressWarnings(library(Biobase)))
suppressPackageStartupMessages(suppressWarnings(library(QDNAseqmod)))
suppressPackageStartupMessages(suppressWarnings(library(cowplot)))
suppressPackageStartupMessages(suppressWarnings(library(ggsignif)))
suppressPackageStartupMessages(suppressWarnings(library(topGO)))
suppressPackageStartupMessages(suppressWarnings(library(org.Hs.eg.db)))
```

### Assign int. functions

```{r functions}
source("data/focal_cna_analysis_funcs.R")
source("../../colour_palettes.R")
```

### Set up directory

```{r setup_dir}
if(!dir.exists("plots/")){
  dir.create("plots/")
}
```

## Load data

### Read meta data

Meta data for all samples with fitted absolute copy number profiles, as well as genomic loci/annotation data for 16 frequently altered genes linked to high grade serous ovarian cancer.

```{r read_meta}
focal_genes_table <- read.table("data/focal_CNA_pseudo_bed.txt",header = T,sep = "\t")
tcga_gene_rates <- read.table("data/TCGA_gene_CNA_rates.txt",header = T,sep = "\t")
tcga_gene_rates <- tcga_gene_rates %>%
                    dplyr::select(-Cytoband) %>%
                    relocate(Gene,Freq,CNA) %>%
                    dplyr::rename(c(gene="Gene",group="CNA",rate="Freq"))

meta.data <- read.table("../../copy_number_signatures/britroc_30kb_signature_data_meta.tsv",header = TRUE,sep = "\t")
patient_data <- read.table("../../britroc_cohort_patient_data.tsv",header = T,sep = "\t")

cytoband <- read.table("data/hg19_cytobands.tsv",header = T,sep = "\t")
cytoband$length <- cytoband$end - cytoband$start
```

### Read absolute data

Absolute copy number profiles are loaded in QDNAseq object format and unused/dropped samples are removed. 

```{r read_abs}
abs_data <- readRDS("../../absolute_POST_down_sampling/britroc_30kb_ds_absCopyNumber.rds")
abs_data <- abs_data[,which(colnames(abs_data) %in% meta.data$SAMPLE_ID[meta.data$use == "TRUE"])]
```

Ploidy as defined by the grid search process for each sample are added to a named vector for use in downstream filtering of copy number states.

```{r set_ploidy}
ploidys <- pData(abs_data)$ploidy
names(ploidys) <- pData(abs_data)$name
```

## Analysis

For the analysis of CNA (both focal/gene-level and broad) copy number thresholds were used as defined by [COSMIC](https://cancer.sanger.ac.uk/cosmic/help/cnv/overview) and [Allele-specific copy number analysis of tumors](https://doi.org/10.1073/pnas.1009843107) publication. 

For focal/gene-level events extending across more than one genome bin, a mean value is taken. 

Broad events are defined on the basis of the proportion of annotated cytoband affected by the CNA. In this case the threshold is 80% of a cytoband called as either AMP or DEL for the entire cytoband to be called as such.

#### Definition of CNA

The copy number alterations, as described above, are defined as such:

##### COSMIC definitions
- Gain:
  average genome ploidy <= 2.7 AND total copy number >= 5
  OR
  average genome ploidy > 2.7 AND total copy number >= 9

- Loss:
  average genome ploidy <= 2.7 AND total copy number = 0
  OR 
  average genome ploidy > 2.7 AND total copy number < ( average genome ploidy - 2.7 )
  
### Focal-level alterations

#### Frequently altered genes in HGSOV

Frequently altered genes are as listed below:

```{r freq_gene_list}
kable(focal_genes_table) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

#### Generate list of protein-coding genes

Using ensembl biomart API, a full list of genes are downloaded and subset to those classified as protein-coding with primary/canonical transcripts, then further limited to autosomes as sex chromosomes are not included in the CNA profile data.

```{r biomart}
if(file.exists("data/protein_coding_genes.tsv")){
	protein_coding_genes <- read.table("data/protein_coding_genes.tsv",header=T,sep="\t")
} else {
	ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl",GRCh = 37)
	protein_coding_genes <- getBM(attributes = c("chromosome_name","start_position","end_position","hgnc_symbol"),
                                             filters = "biotype",
                                             values = "protein_coding",
                                             mart = ensembl,
                                             useCache = FALSE)
	protein_coding_genes <- protein_coding_genes[protein_coding_genes$chromosome_name %in% seq.int(1,22,1),]
	protein_coding_genes <- protein_coding_genes[protein_coding_genes$hgnc_symbol != "",]
	write.table(protein_coding_genes,"data/protein_coding_genes.tsv",col.names=T,row.names=F,sep="\t")
}
kable(head(protein_coding_genes))%>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

#### Extract gene-level CNA

Here quick loading of pre-extracted gene-level CNA values are loaded as this function can take a large amount of time to generate. In an instance where this data is not available a function is executed to tabulate the segment value for each gene across each sample in the BriTROC study.

```{r abs_only_focal}
if(file.exists("data/extracted_output.RDS")){
  extracted <- readRDS("data/extracted_output.RDS")
} else {
  extracted <- extract_abs_segs(abs_data = abs_data,genes = protein_coding_genes) # long execution times
  saveRDS(extracted,file = "data/extracted_output.RDS") 
}
```

```{r saK_data_request}
# request_sample_list <- read.table("../../data_requests/TEMP_SarwahBriTROC_IDs_request.csv",header = T,sep = ",")
# 
# cn_request_table <- extracted[rownames(extracted) %in% focal_genes_table$Gene.name,colnames(extracted) %in% request_sample_list$ID]
# cn_request_table_all <- extracted[,colnames(extracted) %in% request_sample_list$ID]
# write.table(cn_request_table,file = "../../data_requests/cn_genes_abs_dataRequest_SAK.tsv",quote = F,sep = "\t",row.names = T,col.names = T)
# write.table(cn_request_table_all,file = "../../data_requests/cn_all_genes_abs_dataRequest_SAK.tsv",quote = F,sep = "\t",row.names = T,col.names = T)
```


From the gene by sample copy number matrix, copy number gain or losses are called as defined by the criteria described above. The matrix contains three possible values, `N`, `AMP`, and `DEL`.

```{r cna_class_by_ploidy_gene}
cna_calls <- get_cna_calls(data = extracted)
saveRDS(object = cna_calls,file = "data/cna_calls_output.RDS")
```

#### Determining cohort rates - all

From the CNA calls generated from the gene sample copy number matrix, gene-level alteration rates are calculated across the cohort.

##### Specific analysis of genes related to HGSOv 

```{r count_focal_amp}
focal_cna_rates <- get_cna_rates(data = cna_calls,genes = focal_genes_table$Gene.name)

kable(focal_cna_rates[order(focal_cna_rates$rate,decreasing = T),],row.names = F) %>%
  kable_styling(full_width = T) %>%
  scroll_box(height = "200px")
```

### Comparison to TCGA rates

Rates of amplification and deletion between our cohort and TCGA alteration rates are different (with many genes having CNA rates twice that seen in our cohort). This difference is likely due to different sensitivity to CNA calling, in which the parameters used to define CNA alterations in this cohort are more stringent and as such called less frequently.

```{r tcga_gene_rates}
total_amp_rates <- get_cna_rates(data = cna_calls,genes = focal_genes_table$Gene.name)
total_amp_rates$cohort <- rep("britroc",times=nrow(total_amp_rates))
tcga_gene_rates$cohort <- rep("tcga",times=nrow(tcga_gene_rates))

tcga_total_comparison <- rbind(total_amp_rates,tcga_gene_rates)

britroc_tcga_comp_raw <- ggplot(tcga_total_comparison) +
    geom_col(aes(x=gene,y=rate,fill=cohort),position = "dodge") +
    ylab(label = "Rate (%)") +
    facet_wrap(. ~ group) +
    labs(title = "Gene CNA rates between BriTROC and TCGA - raw") +
    theme_bw() + 
    theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90))

saveRDS(object = britroc_tcga_comp_raw,file = "plots/britroc_tcga_comp_raw.RDS")
ggsave(plot = britroc_tcga_comp_raw,filename = "plots/britroc_tcga_comp_raw.png",width = 8,height = 6,units = "in",dpi = 300)
ggsave(plot = britroc_tcga_comp_raw,filename = "plots/britroc_tcga_comp_raw.pdf",width = 8,height = 6,units = "in",dpi = 300)
britroc_tcga_comp_raw
```

Alterations in these genes, although called at lower frequencies, are still called at proportional rates to previous reports, where the most common alterations are consistent between the different data sets.

figure legend - TCGA

Figure XX Amplification and deletion rates between cohorts (TCGA) 
A - Amplification and deletion rates as a percentage of total samples between the BriTROC and the TCGA ovarian serous adenocarcinoma sample set for the sixteen clinically relevant or frequently altered genes listed in supplementary table P6. Rates demonstrate a difference in the total number of focal amplifications and deletions for most genes. 
B - Amplification and deletion rates as described for panel A with a normalisation applied to normalise each rate the highest alteration rate present. The plot demonstrates that though rates are different, the rate at which focal amplifications and deletions are called in the BriTROC cohort are at the same proportions to those seen in TCGA.


```{r tcga_gene_rates_norm}
total_amp_rates_norm <- total_amp_rates
total_amp_rates_norm <- total_amp_rates_norm %>%
  group_by(group) %>%
  mutate(rate=rate / max(rate))

tcga_gene_rates_norm <- tcga_gene_rates
tcga_gene_rates_norm <- tcga_gene_rates_norm %>%
  group_by(group) %>%
  mutate(rate=rate / max(rate))

tcga_total_comparison_norm <- rbind(total_amp_rates_norm,tcga_gene_rates_norm)

britroc_tcga_comp_norm <- ggplot(tcga_total_comparison_norm) +
    geom_col(aes(x=gene,y=rate,fill=cohort),position = "dodge") +
    ylab(label = "normalised rate") +
    facet_wrap(. ~ group) +
    labs(title = "Gene CNA rates between BriTROC and TCGA - normalised") +
    theme_bw() + 
    theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90))

saveRDS(object = britroc_tcga_comp_norm,file = "plots/britroc_tcga_comp_norm.RDS",)
ggsave(plot = britroc_tcga_comp_norm,filename = "plots/britroc_tcga_comp_norm.png",width = 8,height = 6,units = "in",dpi = 300)
ggsave(plot = britroc_tcga_comp_norm,filename = "plots/britroc_tcga_comp_norm.pdf",width = 8,height = 6,units = "in",dpi = 300)
britroc_tcga_comp_norm
```

#### CNA plot - key genes

```{r cna_rate plot,fig.width=12}
p1 <- plot_cna_rates(data = get_cna_rates(data = cna_calls,
                                          genes = focal_genes_table$Gene.name),title = "all")
p2 <- plot_cna_rates(data = get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% meta.data$SAMPLE_ID[meta.data$group == "arx"]],
                                          genes = focal_genes_table$Gene.name),title = "diagnosis")
p3 <- plot_cna_rates(data = get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% meta.data$SAMPLE_ID[meta.data$group == "rlps"]],
                                          genes = focal_genes_table$Gene.name),title = "relapse")

save_plot(filename = "plots/CNA_rates_all_clinrelgenes.png",plot = p1)
save_plot(filename = "plots/CNA_rates_primary_clinrelgenes.png",plot = p2)
save_plot(filename = "plots/CNA_rates_relapse_clinrelgenes.png",plot = p3)
all_rates <- plot_grid(p2 + labs(title = ""),p3 + labs(title = ""),labels = c("A","B"))
ggsave2(filename = "plots/CNA_all_rates.png",plot = all_rates,width = 8,height = 6,units = "in",dpi = 300)
ggsave2(filename = "plots/CNA_all_rates.pdf",plot = all_rates,width = 8,height = 6,units = "in",dpi = 300)
all_rates
```

```{r combined_rates_plot}
arx_rates <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% meta.data$SAMPLE_ID[meta.data$group == "arx"]],
                           genes = focal_genes_table$Gene.name)
arx_rates$type <- rep("diagnosis",times=nrow(arx_rates))
rlps_rates <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% meta.data$SAMPLE_ID[meta.data$group == "rlps"]],
                           genes = focal_genes_table$Gene.name)
rlps_rates$type <- rep("relapse",times=nrow(rlps_rates))

combined_rates <- rbind(arx_rates,rlps_rates)
colnames(combined_rates) <- c("gene","rate","type","group")

combined_rates$rate[combined_rates$type == "DEL"] <- -combined_rates$rate[combined_rates$type == "DEL"]
cp <- ggplot(combined_rates) +
  geom_col(aes(x = gene,y = rate,fill = type),color="grey20") +
  scale_fill_manual(values = colour_palettes$amplification_deletion[c(2,4)]) +
  labs(title = "Focal amplification / deletion rates") +
  xlab("") +
  ylab("rate (%)") +
  theme_bw() +
  theme(axis.title.x = element_blank(),
          axis.text.x = element_text(vjust = 0.5,hjust = 1,angle = 90),
        legend.position = "bottom") +
  facet_wrap(. ~ group)

saveRDS(object = cp,file = "plots/combined_rate_plot_CNA.RDS")
ggsave(filename = "plots/combined_rate_plot_CNA.png",plot = cp,width = 8,height = 6,units = "in",dpi = 300)
ggsave(filename = "plots/combined_rate_plot_CNA.pdf",plot = cp,width = 8,height = 6,units = "in",dpi = 300)
cp
```

```{r compare_gene_rates,fig.width=12}
arx_len <- length(unique(meta.data$SAMPLE_ID[meta.data$group == "arx"]))
rlps_len <- length(unique(meta.data$SAMPLE_ID[meta.data$group == "rlps"]))

arx_rates <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% meta.data$SAMPLE_ID[meta.data$group == "arx"]],
                           genes = focal_genes_table$Gene.name)
arx_rates$type <- rep("diagnosis",times=nrow(arx_rates))

rlps_rates <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% meta.data$SAMPLE_ID[meta.data$group == "rlps"]],
                           genes = focal_genes_table$Gene.name)
rlps_rates$type <- rep("relapse",times=nrow(rlps_rates))

combined_rates <- rbind(arx_rates,rlps_rates)

amp_del_gene_plot <- ggplot(data = combined_rates) +
  geom_col(aes(x = gene,y = rate,fill = type),position = "dodge") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  scale_y_continuous() +
  facet_wrap(. ~ group) +
  theme_bw() +
  theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90))

saveRDS(object = amp_del_gene_plot,file = "plots/amp_del_gene_plot_CNA.RDS",)
ggsave(plot = amp_del_gene_plot,filename = "plots/amp_del_gene_plot_CNA.png",width = 8,height = 6,units = "in",dpi = 300)
ggsave(plot = amp_del_gene_plot,filename = "plots/amp_del_gene_plot_CNA.pdf",width = 8,height = 6,units = "in",dpi = 300)
amp_del_gene_plot
```

### Fishers exact for CNA rates
##### unpaired

```{r cna_rates_fisher_amp}
cna_differences <- fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data$SAMPLE_ID[meta.data$group == "arx"],
                     sub_b = meta.data$SAMPLE_ID[meta.data$group == "rlps"],
                     genes = focal_genes_table$Gene.name)

kable(cna_differences$amplification[order(cna_differences$amplification$p.value) & cna_differences$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences$deletion[order(cna_differences$deletion$p.value) & cna_differences$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

#### paired

```{r cna_rate_paired_fisher}
cna_differences_paired <- fishers_gene_CNA(data = cna_calls,
                            sub_a = meta.data$SAMPLE_ID[meta.data$group == "arx" & meta.data$paired == "TRUE"],
                            sub_b = meta.data$SAMPLE_ID[meta.data$group == "rlps" & meta.data$paired == "TRUE"],
                            genes = focal_genes_table$Gene.name)

kable(cna_differences_paired$amplification[order(cna_differences_paired$amplification$p.value) & cna_differences_paired$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_paired$deletion[order(cna_differences_paired$deletion$p.value) & cna_differences_paired$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

### Prior lines of therapy rate

```{r prior_lines_rates}
table(patient_data$pre_reg_chemo)
meta.data.prior <- meta.data %>%
                    #filter(paired == "TRUE") %>%
                    mutate(prior_lines = patient_data$pre_reg_chemo[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]) %>%
                    mutate(group = case_when(group == "arx" ~ "diagnosis",group == "rlps" ~ "relapse"))

prior_samples <- split(meta.data.prior,f = list(meta.data.prior$prior_lines,meta.data.prior$group))
prior_samples <- prior_samples[lapply(prior_samples,FUN = function(x) nrow(x)) > 1]

prior_rates <- do.call(rbind,lapply(prior_samples,FUN = function(x){
  prior <- meta.data.prior$prior_lines[meta.data.prior$SAMPLE_ID == x$SAMPLE_ID[1]]
  group <- meta.data.prior$group[meta.data.prior$SAMPLE_ID == x$SAMPLE_ID[1]]
  #print(prior)
  samples <- x$SAMPLE_ID
  if(length(samples)>1){
    cnas <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% samples],
                        genes = focal_genes_table$Gene.name)
  } else {
    next()
  }
  cnas$prior_lines <- rep(as.character(prior),times=nrow(cnas))
  cnas$tumour <- rep(as.character(group),times=nrow(cnas))
  return(cnas)
}))

prior_line_plot1 <- ggplot(prior_rates[prior_rates$prior_lines %in% c("1","2","3"),]) +
  geom_col(aes(x = gene,y = rate,fill=tumour),position = "dodge") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  ylab(label = "rate (%)") +
  facet_wrap(group ~ prior_lines,nrow=2,scales = "free_x") + 
    theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90),legend.position = "none")

#prior_lines_cna_focal_plot <- plot_grid(prior_line_plot1,prior_line_plot2,nrow = 2)
ggsave2(filename = "plots/prior_lines_cna_focal_plot.png",plot = prior_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
ggsave2(filename = "plots/prior_lines_cna_focal_plot.pdf",plot = prior_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
prior_line_plot1
```

```{r prior_fishers}
cna_differences_prior1 <- fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "diagnosis" &
                                                         meta.data.prior$prior_lines == "1"],
                     sub_b = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "relapse" &
                                                         meta.data.prior$prior_lines == "1"],
                     genes = focal_genes_table$Gene.name)

cna_differences_prior2 <- fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "diagnosis" &
                                                         meta.data.prior$prior_lines == "2"],
                     sub_b = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "relapse" &
                                                         meta.data.prior$prior_lines == "2"],
                     genes = focal_genes_table$Gene.name)

cna_differences_prior3 <- fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "diagnosis" &
                                                         meta.data.prior$prior_lines == "3"],
                     sub_b = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "relapse" &
                                                         meta.data.prior$prior_lines == "3"],
                     genes = focal_genes_table$Gene.name)
```

```{r prior_fishers_table}
kable(cna_differences_prior1$amplification[order(cna_differences_prior1$amplification$p.value) & cna_differences_prior1$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_prior1$deletion[order(cna_differences_prior1$deletion$p.value) & cna_differences_prior1$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_prior2$amplification[order(cna_differences_prior2$amplification$p.value) & cna_differences_prior2$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_prior2$deletion[order(cna_differences_prior2$deletion$p.value) & cna_differences_prior2$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_prior3$amplification[order(cna_differences_prior3$amplification$p.value) & cna_differences_prior3$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_prior3$deletion[order(cna_differences_prior3$deletion$p.value) & cna_differences_prior3$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

```{r prior_lines_rates_paired}
table(patient_data$pre_reg_chemo)
meta.data.prior <- meta.data %>%
                    filter(paired == "TRUE") %>%
                    mutate(prior_paired_lines = patient_data$pre_reg_chemo[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]) %>%
                    mutate(group = case_when(group == "arx" ~ "diagnosis",group == "rlps" ~ "relapse"))

prior_paired_samples <- split(meta.data.prior,f = list(meta.data.prior$prior_paired_lines,meta.data.prior$group))
prior_paired_samples <- prior_paired_samples[lapply(prior_paired_samples,FUN = function(x) nrow(x)) > 1]

prior_paired_rates <- do.call(rbind,lapply(prior_paired_samples,FUN = function(x){
  prior <- meta.data.prior$prior_paired_lines[meta.data.prior$SAMPLE_ID == x$SAMPLE_ID[1]]
  group <- meta.data.prior$group[meta.data.prior$SAMPLE_ID == x$SAMPLE_ID[1]]
  #print(prior)
  samples <- x$SAMPLE_ID
  if(length(samples)>1){
    cnas <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% samples],
                        genes = focal_genes_table$Gene.name)
  } else {
    next()
  }
  cnas$prior_paired_lines <- rep(as.character(prior),times=nrow(cnas))
  cnas$tumour <- rep(as.character(group),times=nrow(cnas))
  return(cnas)
}))

prior_paired_line_plot1 <- ggplot(prior_paired_rates) +
  geom_col(aes(x = gene,y = rate,fill=tumour),position = "dodge") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  ylab(label = "rate (%)") +
  facet_wrap(prior_paired_lines ~ group,nrow=2,scales = "free_x") + 
    theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90),legend.position = "none")

#prior_paired_lines_cna_focal_plot <- plot_grid(prior_paired_line_plot1,prior_paired_line_plot2,nrow = 2)
ggsave2(filename = "plots/prior_paired_lines_cna_focal_paired_plot.png",plot = prior_paired_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
ggsave2(filename = "plots/prior_paired_lines_cna_focal_paired_plot.pdf",plot = prior_paired_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
prior_paired_line_plot1
```

```{r prior_fishers_paired}
cna_differences_paired_prior1 <- fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "diagnosis" &
                                                         meta.data.prior$prior_paired_lines == "1" &
                                                         meta.data.prior$paired == "TRUE"],
                     sub_b = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "relapse" &
                                                         meta.data.prior$prior_paired_lines == "1" &
                                                         meta.data.prior$paired == "TRUE"],
                     genes = focal_genes_table$Gene.name)

cna_differences_paired_prior2 <- fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "diagnosis" &
                                                         meta.data.prior$prior_paired_lines == "2" &
                                                         meta.data.prior$paired == "TRUE"],
                     sub_b = meta.data.prior$SAMPLE_ID[meta.data.prior$group == "relapse" &
                                                         meta.data.prior$prior_paired_lines == "2" &
                                                         meta.data.prior$paired == "TRUE"],
                     genes = focal_genes_table$Gene.name)
```

```{r prior_fishers_paired_table}
kable(cna_differences_paired_prior1$amplification[order(cna_differences_paired_prior1$amplification$p.value) &
                                                    cna_differences_paired_prior1$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_paired_prior1$deletion[order(cna_differences_paired_prior1$deletion$p.value) &
                                               cna_differences_paired_prior1$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_paired_prior2$amplification[order(cna_differences_paired_prior2$amplification$p.value) &
                                                    cna_differences_paired_prior2$amplification$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")

kable(cna_differences_paired_prior2$deletion[order(cna_differences_paired_prior2$deletion$p.value) &
                                               cna_differences_paired_prior2$deletion$p.value < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

```{r prior_plot_comb}
prior_lines_cna_focal_plot <- plot_grid(prior_line_plot1,prior_paired_line_plot1,nrow = 2,labels = c("A","B"))
ggsave2(plot = prior_lines_cna_focal_plot,
        filename = "plots/prior_paired_lines_combined_plot.png",
        height = 8,width = 10,units = "in",dpi = 300)
ggsave2(plot = prior_lines_cna_focal_plot,
        filename = "plots/prior_paired_lines_combined_plot.pdf",
        height = 8,width = 10,units = "in",dpi = 300)

prior_lines_cna_focal_plot
```

### platinum status rate

```{r pt_status_rates}
table(patient_data$pt_sensitivity_at_reg,patient_data$pt_sensitivity_at_reg)
meta.data.pt <- meta.data %>%
                    #filter(paired == "TRUE") %>%
                    mutate(pt_status = patient_data$pt_sensitivity_at_reg[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]) %>%
                    mutate(group = case_when(group == "arx" ~ "diagnosis",group == "rlps" ~ "relapse"))

pt_samples <- split(meta.data.pt,f = list(meta.data.pt$pt_status,meta.data.pt$group))
pt_samples <- pt_samples[lapply(pt_samples,FUN = function(x) nrow(x)) > 1]

pt_rates <- do.call(rbind,lapply(pt_samples,FUN = function(x){
  pt <- meta.data.pt$pt_status[meta.data.pt$SAMPLE_ID == x$SAMPLE_ID[1]]
  group <- meta.data.pt$group[meta.data.pt$SAMPLE_ID == x$SAMPLE_ID[1]]
  #print(pt)
  samples <- x$SAMPLE_ID
  if(length(samples)>1){
    cnas <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% samples],
                        genes = focal_genes_table$Gene.name)
  } else {
    next()
  }
  cnas$pt_status <- rep(as.character(pt),times=nrow(cnas))
  cnas$tumour <- rep(as.character(group),times=nrow(cnas))
  return(cnas)
}))

pt_line_plot1 <- ggplot(pt_rates) +
  geom_col(aes(x = gene,y = rate,fill=tumour),position = "dodge") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  ylab(label = "rate (%)") +
  facet_wrap(pt_status ~ group,nrow=2,scales = "free_x") + 
    theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90),legend.position = "none")

#pt_status_cna_focal_plot <- plot_grid(pt_line_plot1,pt_line_plot2,nrow = 2)
ggsave2(filename = "plots/pt_status_cna_focal_plot.png",plot = pt_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
ggsave2(filename = "plots/pt_status_cna_focal_plot.pdf",plot = pt_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
pt_line_plot1
```

```{r pt_status_fishers}
fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "diagnosis" &
                                                         meta.data.pt$pt_status == "sensitive"],
                     sub_b = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "relapse" &
                                                         meta.data.pt$pt_status == "sensitive"],
                     genes = focal_genes_table$Gene.name)

fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "diagnosis" &
                                                         meta.data.pt$pt_status == "resistant"],
                     sub_b = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "relapse" &
                                                         meta.data.pt$pt_status == "resistant"],
                     genes = focal_genes_table$Gene.name)
```

```{r pt_status_rates_paired}
table(patient_data$pt_sensitivity_at_reg)
meta.data.pt <- meta.data %>%
                    filter(paired == "TRUE") %>%
                    mutate(pt_paired_lines = patient_data$pt_sensitivity_at_reg[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]) %>%
                    mutate(group = case_when(group == "arx" ~ "diagnosis",group == "rlps" ~ "relapse"))

pt_paired_samples <- split(meta.data.pt,f = list(meta.data.pt$pt_paired_lines,meta.data.pt$group))
pt_paired_samples <- pt_paired_samples[lapply(pt_paired_samples,FUN = function(x) nrow(x)) > 1]

pt_paired_rates <- do.call(rbind,lapply(pt_paired_samples,FUN = function(x){
  pt <- meta.data.pt$pt_paired_lines[meta.data.pt$SAMPLE_ID == x$SAMPLE_ID[1]]
  group <- meta.data.pt$group[meta.data.pt$SAMPLE_ID == x$SAMPLE_ID[1]]
  #print(pt)
  samples <- x$SAMPLE_ID
  if(length(samples)>1){
    cnas <- get_cna_rates(data = cna_calls[,colnames(cna_calls) %in% samples],
                        genes = focal_genes_table$Gene.name)
  } else {
    next()
  }
  cnas$pt_paired_lines <- rep(as.character(pt),times=nrow(cnas))
  cnas$tumour <- rep(as.character(group),times=nrow(cnas))
  return(cnas)
}))

pt_paired_line_plot1 <- ggplot(pt_paired_rates) +
  geom_col(aes(x = gene,y = rate,fill=tumour),position = "dodge") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  ylab(label = "rate (%)") +
  facet_wrap(pt_paired_lines ~ group,nrow=2,scales = "free_x") + 
    theme(axis.title.x = element_blank(),axis.text.x = element_text(vjust = 0.5,angle = 90),legend.position = "none")

#pt_paired_lines_cna_focal_plot <- plot_grid(pt_paired_line_plot1,pt_paired_line_plot2,nrow = 2)
ggsave2(filename = "plots/pt_paired_cna_focal_paired_plot.png",plot = pt_paired_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
ggsave2(filename = "plots/pt_paired_cna_focal_paired_plot.pdf",plot = pt_paired_line_plot1,width = 8,height = 8,units = "in",dpi = 300)
pt_paired_line_plot1
```

```{r pt_status_fishers_paired}
print("sensitive-paired")
fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "diagnosis" &
                                                         meta.data.pt$paired == "TRUE" &
                                                         meta.data.pt$pt_paired_lines == "sensitive"],
                     sub_b = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "relapse" &
                                                         meta.data.pt$paired == "TRUE" &
                                                         meta.data.pt$pt_paired_lines == "sensitive"],
                     genes = focal_genes_table$Gene.name)
print("resistant-paired")
fishers_gene_CNA(data = cna_calls,
                     sub_a = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "diagnosis" &
                                                         meta.data.pt$paired == "TRUE" &
                                                         meta.data.pt$pt_paired_lines == "resistant"],
                     sub_b = meta.data.pt$SAMPLE_ID[meta.data.pt$group == "relapse" &
                                                         meta.data.pt$paired == "TRUE" &
                                                         meta.data.pt$pt_paired_lines == "resistant"],
                     genes = focal_genes_table$Gene.name)
```

```{r pt_plot_comb}
pt_status_cna_focal_plot <- plot_grid(pt_line_plot1,pt_paired_line_plot1,nrow = 2,labels = c("A","B"))
ggsave2(plot = pt_status_cna_focal_plot,
        filename = "plots/pt_paired_combined_plot.png",
        height = 8,width = 10,units = "in",dpi = 300)
ggsave2(plot = pt_status_cna_focal_plot,
        filename = "plots/pt_paired_combined_plot.pdf",
        height = 8,width = 10,units = "in",dpi = 300)

pt_status_cna_focal_plot
```

### Gene copy counts

Comparison of CNA rates between samples is a single metric to measure the prevalence of CNA changes between samples but as this study has absolute copy number profiles, an alternative is to measure copy number value distributions and net changes between diagnosis and relapse settings.

```{r copies_ext}
copies <- extracted
```

#### Mann-Whitney amplification magnitude - paired

Using the changes derived above, a mann-whitney can be used across the cohort to test the median amplification state of a given gene to test for a coordinated shift in CNA between primary and relapse cases. Below performs this test in both the paired and all samples.

For specific genes of interest, most genes with sufficient amplification observations did not demonstrate a coordinated shift in copy number state, for both paired and unpaired sample sets.

```{r gene_copies_paired}
cna_copies_results_paired <- cna_median_copies(data = copies,min_obs = 3,
                         sub_a = meta.data$SAMPLE_ID[meta.data$group == "arx" & meta.data$paired == TRUE],
                         sub_b = meta.data$SAMPLE_ID[meta.data$group == "rlps" & meta.data$paired == TRUE],p.adj = "fdr")

kable(cna_copies_results_paired[cna_copies_results_paired$pVal %in% focal_genes_table$Gene.name,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```


```{r copies_paired_all_genes}
kable(cna_copies_results_paired[cna_copies_results_paired$pVal < 0.05,]) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

### Copy changes

Individual patients show distinct shifts in amplification rates for genes of clinical interest, where amplification rates can either remain constant, reduce, or increase, both collectively and independently. Plots below show the range of profile changes between primary and relapse. 

```{r patient_level_changes,fig.width=12}
paired_amp_changes_all <- get_patient_changes(data = copies,
                                              genes = protein_coding_genes$hgnc_symbol,
                                              patients = unique(meta.data$PATIENT_ID[meta.data$paired == TRUE])) %>%
                       mutate(group = case_when(group == "rlps" ~ "relapse",group == "arx" ~ "diagnosis")) %>%
                       mutate(pt_status = patient_data$pt_sensitivity_at_reg[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]) %>%
                       mutate(prior_lines = as.factor(patient_data$pre_reg_chemo[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]))

paired_amp_changes <- paired_amp_changes_all %>%
                        filter(gene %in% focal_genes_table$Gene.name)
                        
copy_state_groups <- ggplot(paired_amp_changes) +
  #geom_point(aes(x = gene,y=value,fill=group),position="jitter",alpha = 0.3) +
  geom_line(aes(x = group,y=value,group=PATIENT_ID),alpha = 0.3) +
  geom_point(aes(x = group,y=value,color=group),
             position=position_jitterdodge(jitter.width = 0.2),
             alpha=0.2) +
  geom_boxplot(aes(x = group,y=value,fill=group),outlier.shape = NA) +
  ylab("copy number") +
  xlab("") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  scale_color_manual(values = colour_palettes$diagnosis_relapse) +
  facet_wrap(. ~ gene,scales = "free") +
  ggsignif::geom_signif(aes(x = group,y=value),vjust = 1.2,
                        comparisons = list(c("diagnosis","relapse")),
                        test = "wilcox.test") +
  theme_bw() +
  theme(legend.position = "bottom")

saveRDS(object = copy_state_groups,file = "plots/copy_state_groups.RDS")
ggsave(plot = copy_state_groups,filename = "plots/copy_state_groups.png",width = 10,height = 8,units = "in",dpi = 300)
copy_state_groups

pats <- unique(meta.data$PATIENT_ID[meta.data$paired == TRUE])[12:16]
paired_amp_changes_plot <- ggplot(paired_amp_changes[paired_amp_changes$PATIENT_ID %in% pats,]) +
  geom_point(aes(x=gene,y=value,color=group)) +
  scale_color_manual(values = colour_palettes$diagnosis_relapse) +
  coord_flip() +
  facet_wrap(. ~ PATIENT_ID,
             nrow = ceiling(sqrt(length(unique(paired_amp_changes$PATIENT_ID)))),
             ncol = ceiling(sqrt(length(unique(paired_amp_changes$PATIENT_ID)))),
             scales = "free") +
  theme_bw()

ggsave(plot = paired_amp_changes_plot,filename = "plots/paired_amp_changes_plot.png",width = 8,height = 6,units = "in",dpi = 300)
ggsave(plot = paired_amp_changes_plot,filename = "plots/paired_amp_changes_plot.pdf",width = 8,height = 6,units = "in",dpi = 300)
paired_amp_changes_plot
```

### Paired prior lines of therapy copies

```{r paired_prior_lines_copies}
paired_gene_prior_lines <- ggplot(paired_amp_changes[paired_amp_changes$prior_lines %in% c(1,2),]) +
  geom_point(aes(x = group,y = value),alpha=0.2,position = position_dodge2(width = 0.4)) +
  geom_boxplot(aes(x = group,y = value,fill=group),alpha=0.3,position = "dodge") +
  geom_signif(aes(x = group,y = value),comparison=list(c("diagnosis","relapse")),vjust = 1.5,test = "wilcox.test") +
  facet_wrap(as.factor(prior_lines) ~ gene,scales = "free_y") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.text.x = element_text(angle = 90),
        axis.title.x = element_blank())

table(paired_amp_changes$prior_lines[paired_amp_changes$prior_lines %in% c(1,2) & paired_amp_changes$gene == "AKT1"],
      paired_amp_changes$group[paired_amp_changes$prior_lines %in% c(1,2) & paired_amp_changes$gene == "AKT1"])

ggsave(plot = paired_gene_prior_lines,filename = "plots/paired_gene_prior_lines_plot.png",width = 6,height = 8,units = "in",dpi = 300)
ggsave(plot = paired_gene_prior_lines,filename = "plots/paired_gene_prior_lines_plot.pdf",width = 6,height = 8)
paired_gene_prior_lines
```

```{r paired_prior_lines_testing}
cn_copies_prior_1 <- cna_median_copies(data = copies,min_obs = 3,
                         sub_a = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "diagnosis" & paired_amp_changes$prior_lines == "1"],
                         sub_b = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "relapse" & paired_amp_changes$prior_lines == "1"],p.adj = "fdr")

cn_copies_prior_2 <- cna_median_copies(data = copies,min_obs = 3,
                         sub_a = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "diagnosis" & paired_amp_changes$prior_lines == "2"],
                         sub_b = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "relapse" & paired_amp_changes$prior_lines == "2"],p.adj = "fdr")

head(cn_copies_prior_1)
head(cn_copies_prior_2)
```

### Paired Pt status copies

```{r paired_pt_copies}
paired_gene_pt_status <- ggplot(paired_amp_changes) +
  geom_jitter(aes(x = group,y = value),alpha=0.2,position = position_dodge2(width = 0.4)) +
  geom_boxplot(aes(x = group,y = value,fill=group),alpha=0.3,position = "dodge") +
  geom_signif(aes(x = group,y = value),comparison=list(c("diagnosis","relapse")),vjust = 1.5,test = "wilcox.test") +
  facet_wrap(pt_status ~ gene,scales = "free_y") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 90))

table(paired_amp_changes$pt_status[paired_amp_changes$gene == "AKT1"],
      paired_amp_changes$group[paired_amp_changes$gene == "AKT1"])

ggsave(plot = paired_gene_pt_status,filename = "plots/paired_gene_pt_status_plot.png",width = 6,height = 8,units = "in",dpi = 300)
ggsave(plot = paired_gene_pt_status,filename = "plots/paired_gene_pt_status_plot.pdf",width = 6,height = 8)
paired_gene_pt_status
```

```{r paired_pt_status_testing}
cn_copies_pt_sensitive <- cna_median_copies(data = copies,min_obs = 3,
                         sub_a = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "diagnosis" & paired_amp_changes$pt_status == "sensitive"],
                         sub_b = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "relapse" & paired_amp_changes$pt_status == "sensitive"],
                         p.adj = "fdr")

cn_copies_pt_resist <- cna_median_copies(data = copies,min_obs = 3,
                         sub_a = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "diagnosis" & paired_amp_changes$pt_status == "resistant"],
                         sub_b = paired_amp_changes$SAMPLE_ID[paired_amp_changes$group == "relapse" & paired_amp_changes$pt_status == "resistant"],
                         p.adj = "fdr")
cn_copies_pt_sensitive[cn_copies_pt_sensitive$qVal < 0.05,]
write.table(x = cn_copies_pt_sensitive[cn_copies_pt_sensitive$qVal < 0.05,],
            file = "data/cn_copies_pt_sensitive.tsv",quote = F,sep = "\t",row.names = F,col.names = T)
head(cn_copies_pt_resist)
```

```{r paired_pt_status_enrich}
GO.ids <- mapIds(x = org.Hs.eg.db,keys = as.character(cn_copies_pt_sensitive$Gene),keytype = "SYMBOL",column = "GO",multiVals = "list")
geneL <- cn_copies_pt_sensitive$qVal
names(geneL) <- as.character(cn_copies_pt_sensitive$Gene)
GOdata <- new("topGOdata",
              ontology = "BP",
              allGenes = geneL,
              geneSel = geneSelection,
              annot = annFUN.gene2GO,
              gene2GO = GO.ids,
              nodeSize = 5)

resultFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
allRes <- GenTable(GOdata,Fisher = resultFisher,topNodes = 500,numChar = 5000)
allRes$Fisher <- as.numeric(allRes$Fisher)
allRes$qVal <- p.adjust(allRes$Fisher,method = "fdr")

kable(allRes[allRes$qVal < 0.05,])  %>%
  kable_styling() %>%
  scroll_box(height = "200px")

write.table(x = allRes[allRes$qVal < 0.05,],file = "data/pt_sensitive_topGO_results.tsv",quote = F,sep = "\t",row.names = F,col.names = T)
```

#### Normalised changes

The profiles from above are dependent on ploidy value and distance between primary and relapse are more difficult to interpret in this context, so normalising the divergence between primary and relapse, where the median primary value is set to 0 allows for easier interpretation.

```{r norm_patiet_level_changes,fig.width=12}
norm_amp_changes_all <- paired_amp_changes_all %>%
  group_by(PATIENT_ID,group,gene) %>%
  summarise(value=median(value)) %>%
  pivot_wider(values_from = value, names_from=group) %>%
  mutate(change=relapse-diagnosis)

norm_amp_changes <- norm_amp_changes_all %>%
                      filter(gene %in% focal_genes_table$Gene.name)
ylim <- max(abs(norm_amp_changes$change[norm_amp_changes$PATIENT_ID %in% pats]))
#ylim <- max(abs(norm_amp_changes$change))
ggplot(norm_amp_changes[norm_amp_changes$PATIENT_ID %in% pats,]) +
  geom_point(aes(x=gene,y=change)) +
  geom_vline(xintercept = 0) +
  scale_y_continuous(limits = c(-ylim*1.05,ylim*1.05)) +
  coord_flip() +
  facet_wrap(. ~ PATIENT_ID,
             nrow = ceiling(sqrt(length(unique(paired_amp_changes$PATIENT_ID)))),
             ncol = ceiling(sqrt(length(unique(paired_amp_changes$PATIENT_ID)))),
             scales = "free") +
  theme_bw()
```

### all changes + normalisation

```{r paired_all_change_diff_plot}
norm_amp_changes_all_median <- norm_amp_changes_all %>%
                                dplyr::select(-diagnosis,-relapse) %>%
                                group_by(gene) %>%
                                summarise(across(change,median)) %>%
                                ungroup() %>%
                                mutate(pval = cna_copies_results_paired$pVal[match(gene,cna_copies_results_paired$Gene)]) %>%
                                mutate(qval = cna_copies_results_paired$qVal[match(gene,cna_copies_results_paired$Gene)]) %>%
                                mutate(direction = case_when(pval < 0.05 & change > 0 ~ "gain",
                                                             pval < 0.05 & change < 0 ~ "loss",
                                                             pval > 0.05 ~ "neutral"))

head(norm_amp_changes_all_median[order(norm_amp_changes_all_median$pval),])

ggplot(norm_amp_changes_all_median) +
  geom_point(aes(change,-log10(pval),color=direction),
             alpha = 0.1) +
  geom_hline(yintercept = -log10(0.05),linetype=2) +
  geom_hline(yintercept = -log10(0.000002769623),linetype=2) + # FDR correction @ 18,053
  scale_x_continuous(limits = c(-1,1)) +
  scale_color_manual(values = c(colour_palettes$amp_del[c(1,3)],"neural"="grey80")) +
  theme_bw() +
  annotate(geom = "text",x = 1,y = -log10(0.065),label = "p.val") +
  annotate(geom = "text",x = 1,y = -log10(0.000004),label = "q.val")
```

### Prior lines of therapy copies

```{r prior_lines_copies}
meta.data.prior <- meta.data %>%
                    filter(paired == "TRUE") %>%
                    mutate(prior_lines = patient_data$pre_reg_chemo[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))])

norm_amp_changes$prior_lines <- patient_data$pre_reg_chemo[match(norm_amp_changes$PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]

norm_amp_changes$prior_lines <- as.factor(norm_amp_changes$prior_lines)

gene_change_prior_lines <- ggplot(norm_amp_changes[norm_amp_changes$prior_lines %in% c(1,2),]) +
  geom_point(aes(x = prior_lines,y = change,color=prior_lines),position = position_dodge2(width = 0.4)) +
  geom_boxplot(aes(x = prior_lines,y = change,fill=prior_lines),alpha=0.3,position = "dodge") +
  geom_signif(aes(x = prior_lines,y = change),comparison=list(c("1","2")),test = "wilcox.test",vjust = 1.5) +
  xlab("") +
  scale_fill_manual(values=c("#e5f5f9","#99d8c9"),name = "prior lines") +
  scale_color_manual(values=c("#e5f5f9","#99d8c9"),name = "prior lines") +
  facet_wrap(. ~ gene,scales = "free_y",ncol = 3) +
  theme_bw() +
  theme(legend.position = "bottom")

ggsave(plot = gene_change_prior_lines,filename = "plots/gene_change_prior_lines_plot.png",width = 6,height = 8,units = "in",dpi = 300)
ggsave(plot = gene_change_prior_lines,filename = "plots/gene_change_prior_lines_plot.pdf",width = 6,height = 8)
gene_change_prior_lines
```

### Pt status copies

```{r test_changes_pt_status_paired}
norm_amp_changes_wide <- norm_amp_changes_all %>%
  dplyr::select(PATIENT_ID,gene,change) %>%
  pivot_wider(id_cols = gene,names_from = PATIENT_ID,values_from = change) %>%
  column_to_rownames(var = "gene")

meta.data.prior <- meta.data %>%
                    filter(paired == "TRUE") %>%
                    mutate(pt_status = patient_data$pre_reg_chemo[match(PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))])

norm_amp_changes$pt_status <- patient_data$pt_sensitivity_at_reg[match(norm_amp_changes$PATIENT_ID,paste0("BRITROC-",patient_data$britroc_number))]

norm_amp_changes$pt_status <- as.factor(norm_amp_changes$pt_status)

cn_copies_pt_status <- cna_median_copies(data = norm_amp_changes_wide,min_obs = 3,
                         sub_a = unique(norm_amp_changes$PATIENT_ID[norm_amp_changes$pt_status == "resistant"]),
                         sub_b = unique(norm_amp_changes$PATIENT_ID[norm_amp_changes$pt_status == "sensitive"]),p.adj = "fdr")

kable(cn_copies_pt_status) %>%
  kable_styling() %>% scroll_box(height = "200px")

cn_copies_pt_status[cn_copies_pt_status$Gene %in% focal_genes_table$Gene.name,]
```

```{r pt_copies}
gene_change_pt_status <- ggplot(norm_amp_changes) +
  geom_point(aes(x = pt_status,y = change,color=pt_status),position = position_dodge2(width = 0.4)) +
  geom_boxplot(aes(x = pt_status,y = change,fill=pt_status),alpha=0.3,position = "dodge") +
  geom_signif(aes(x = pt_status,y = change),comparison=list(c("resistant","sensitive")),test = "wilcox.test",vjust = 1.5) +
  facet_wrap(. ~ gene,scales = "free_y",ncol = 3) +
  theme_bw() +
  theme(legend.position = "bottom")

ggsave(plot = gene_change_pt_status,filename = "plots/gene_change_pt_status_plot.png",width = 6,height = 8,units = "in",dpi = 300)
ggsave(plot = gene_change_pt_status,filename = "plots/gene_change_pt_status_plot.pdf",width = 6,height = 8)
gene_change_pt_status
```

```{r sik2}
fishers_gene_CNA(data = cna_calls,
                            sub_a = meta.data$SAMPLE_ID[meta.data$group == "arx" & meta.data$paired == "TRUE"],
                            sub_b = meta.data$SAMPLE_ID[meta.data$group == "rlps" & meta.data$paired == "TRUE"],
                            genes = "SIK2")

cna_copies_results_paired[cna_copies_results_paired$Gene == "SIK2",]

ggplot(paired_amp_changes_all[paired_amp_changes_all$gene == "SIK2",]) +
  #geom_point(aes(x = gene,y=value,fill=group),position="jitter",alpha = 0.3) +
  geom_line(aes(x = group,y=value,group=PATIENT_ID),alpha = 0.3) +
  geom_point(aes(x = group,y=value,color=group),
             position=position_jitterdodge(jitter.width = 0.2),
             alpha=0.2) +
  geom_boxplot(aes(x = group,y=value,fill=group),outlier.shape = NA) +
  ylab("copy number") +
  xlab("") +
  scale_fill_manual(values = colour_palettes$diagnosis_relapse) +
  scale_color_manual(values = colour_palettes$diagnosis_relapse) +
  facet_wrap(. ~ gene,scales = "free") +
  ggsignif::geom_signif(aes(x = group,y=value),vjust = 1.2,
                        comparisons = list(c("diagnosis","relapse")),
                        test = "wilcox.test") +
  theme_bw() +
  theme(legend.position = "bottom")
```
## Session info

```{r session_info}
sessionInfo()
```
