---
title: "Analysis of absolute CNAs - BriTROC 30kb - focal (patient specific)"
author: "Philip Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: TRUE
    toc_float: TRUE
---

## Initial steps

### knitr options

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries

```{r load_libraries}
# Load libraries
suppressPackageStartupMessages(suppressWarnings(library(Biobase)))
suppressPackageStartupMessages(suppressWarnings(library(QDNAseqmod)))
suppressPackageStartupMessages(suppressWarnings(library(tidyverse)))
suppressPackageStartupMessages(suppressWarnings(library(umap)))
suppressPackageStartupMessages(suppressWarnings(library(cowplot)))
suppressPackageStartupMessages(suppressWarnings(library(kableExtra)))
suppressPackageStartupMessages(suppressWarnings(library(biomaRt)))
suppressPackageStartupMessages(suppressWarnings(library(ComplexHeatmap)))
#suppressPackageStartupMessages(suppressWarnings(library(topGO)))
suppressPackageStartupMessages(suppressWarnings(library(VennDiagram)))
suppressPackageStartupMessages(suppressWarnings(library(ggsignif)))
suppressPackageStartupMessages(suppressWarnings(library(RColorBrewer)))

```

### Assign int. functions

```{r functions}
source("data/focal_cna_analysis_funcs.R")
source("data/nbclustfixed.R")
source("../../colour_palettes.R")
```

## Load data

### Read meta data

Meta data for all samples with fitted absolute copy number profiles, as well as genomic loci/annotation data for 16 frequently altered genes linked to high grade serous ovarian cancer.

```{r read_meta}
meta.data <- read.table("../../copy_number_signatures/britroc_30kb_signature_data_meta.tsv",header = TRUE,sep = "\t")
patient_data <- read.table("../../britroc_cohort_patient_data.tsv",header = T,sep = "\t")

cytoband <- read.table("data/hg19_cytobands.tsv",header = T,sep = "\t")
cytoband$length <- cytoband$end - cytoband$start

ploidy_change_samples <- read.table(file = "../broad_analysis/data/ploidy_change_pats.tsv",header = T,sep = "\t")
ploidy_change_samples <- unique(ploidy_change_samples$PATIENT_ID)

#signatures
sig.exposures <- read.table("../../copy_number_signatures/britroc_30kb_signature_data_sig_exposures.tsv",header = TRUE,sep = "\t")

#ith
ith <- read.table("../heterogeneity/data/ith_results_table.tsv",header = TRUE,sep = "\t")

# BRCA
```

Frequently altered genes are as listed below:

```{r load_genes}
focal_genes_table <- read.table("data/focal_CNA_pseudo_bed.txt",header = T,sep = "\t")
tcga_gene_rates <- read.table("data/TCGA_gene_CNA_rates.txt",header = T,sep = "\t")
tcga_gene_rates <- tcga_gene_rates[,-2]
tcga_gene_rates <- tcga_gene_rates[,c(1,3,2)]
colnames(tcga_gene_rates) <- c("gene","rate","group")
tcga_gene_rates$group <- gsub(pattern = "HOMDEL",replacement = "DEL",x = tcga_gene_rates$group)
```

```{r freq_gene_list}
kable(focal_genes_table) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

### Read absolute data

Absolute copy number profiles are loaded in QDNAseq object format and unused/dropped samples are removed. 

```{r read_abs}
abs_data <- readRDS("../../absolute_POST_down_sampling/britroc_30kb_ds_absCopyNumber.rds")
abs_data <- abs_data[,which(colnames(abs_data) %in% meta.data$SAMPLE_ID[meta.data$use == "TRUE"])]
```

Ploidy as defined by the grid search process for each sample are added to a named vector for use in downstream filtering of copy number states.

```{r set_ploidy}
ploidys <- pData(abs_data)$ploidy
names(ploidys) <- pData(abs_data)$name
```

Using ensembl biomart API, a full list of genes are downloaded and subset to those classified as protein-coding with primary/canonical transcripts, then further limited to autosomes as sex chromosomes are not included in the CNA profile data.

```{r biomart}
ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl",GRCh = 37)
protein_coding_genes <- getBM(attributes = c("chromosome_name","start_position","end_position","hgnc_symbol"),
                                             filters = "biotype",
                                             values = "protein_coding",
                                             mart = ensembl,
                                             useCache = FALSE)
protein_coding_genes <- protein_coding_genes[protein_coding_genes$chromosome_name %in% seq.int(1,22,1),]
protein_coding_genes <- protein_coding_genes[protein_coding_genes$hgnc_symbol != "",]
protein_coding_genes$chromosome_name <- as.numeric(protein_coding_genes$chromosome_name)
protein_coding_genes <- protein_coding_genes[order(protein_coding_genes$chromosome_name,protein_coding_genes$start_position),]
```

#### Extract gene-level CNA

Here quick loading of pre-extracted gene-level CNA values are loaded as this function can take a large amount of time to generate. In an instance where this data is not available a function is executed to tabulate the segment value for each gene across each sample in the BriTROC study.

```{r abs_only_focal}
if(file.exists("data/extracted_output.RDS")){
  extracted <- readRDS("data/extracted_output.RDS")
} else {
  extracted <- extract_abs_segs(abs_data = abs_data,genes = protein_coding_genes) # long execution times
  saveRDS(extracted,file = "data/extracted_output.RDS") 
}
```

```{r copies_ext}
copies <- extracted
```

```{r patient_level_changes,fig.width=12}
paired_changes <- get_patient_changes(data = copies,
                                              genes = focal_genes_table$Gene.name,
                                              patients = unique(meta.data$PATIENT_ID[meta.data$paired == TRUE]))

norm_paired_changes <- paired_changes %>%
  group_by(PATIENT_ID,group,gene) %>%
  summarise(value=median(value)) %>%
  pivot_wider(values_from = value, names_from=group) %>%
  mutate(change=rlps-arx)
```

### Patient-specific changes

```{r pat_specific_justification}
pat_ordering <- norm_paired_changes %>%
                  group_by(PATIENT_ID) %>%
                  summarise(across(change,sum)) %>%
                  arrange(change)

norm_paired_changes$PATIENT_ID <- factor(norm_paired_changes$PATIENT_ID,
                                         levels = as.character(rev(pat_ordering$PATIENT_ID)))

ggplot(norm_paired_changes) +
  geom_hline(yintercept = 0) +
  geom_col(aes(x = PATIENT_ID,y = change,group = gene,fill=ifelse(PATIENT_ID %in% ploidy_change_samples,"changed","unchanged"))) +
  facet_wrap(. ~ gene,scales = "free_y",ncol = 3) +
  ylab("copy number change") +
  xlab("patient") +
  scale_fill_manual(name = "ploidy change",values = c("firebrick2","grey50")) +
  theme_bw() +
  theme(axis.text.x = element_blank(),axis.ticks = element_blank())
```

```{r normalised_changes_stats_genelist}
norm_paired_change_summary <- norm_paired_changes %>%
                        mutate(shift = ifelse(change < 0, "loss", "gain")) %>%
                        group_by(shift) %>%
                        dplyr::select(-gene)

norm_change_summary_median <- norm_paired_change_summary %>%
                                group_by(PATIENT_ID) %>%
                                summarise(median=median(change)) %>%
                                arrange(median)

norm_paired_change_summary$PATIENT_ID <- factor(norm_paired_change_summary$PATIENT_ID,levels = unique(norm_change_summary_median$PATIENT_ID))

ggplot(norm_paired_change_summary) +
  geom_point(aes(change,PATIENT_ID,color=shift),alpha = 0.5) +
  geom_boxplot(aes(change,PATIENT_ID,fill=shift),outlier.alpha = 0)
```

```{r change_cor_mat_select_genes}
cor_data <- norm_paired_changes %>%
  filter(!PATIENT_ID %in% ploidy_change_samples) %>%
  dplyr::select(-c(arx,rlps)) %>%
  pivot_wider(id_cols = PATIENT_ID,names_from = gene,values_from = change) %>%
  column_to_rownames(var = "PATIENT_ID")

cor_matrix <- Hmisc::rcorr(x = as.matrix(cor_data),type = "spearman")

plot_colours <- colorRampPalette(rev(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA")))
corrplot::corrplot(cor_matrix$r,col = col(matrix(200)),mar = c(1, 0, 1, 0),
                   method = "color",
                   p.mat = cor_matrix$P,
                   addCoef.col = "black",
                   number.cex = 0.7,
                   tl.col = "black",
                   tl.srt = 90,
                   tl.cex = 0.7,
                   type = "upper",
                   sig.level = 0.01,
                   insig = "blank",order = "hclust",
                   diag = FALSE)

p1_formula <- ~{
  corrplot::corrplot(corr = cor_matrix$r,number.digits = 2,col = plot_colours(200),
                   mar = c(1, 0, 1, 0),
                   method = "color",addgrid.col = "grey90",
                   p.mat = cor_matrix$P,
                   addCoef.col = "black",
                   number.cex = 0.6,
                   tl.col = "black",
                   tl.srt = 90,
                   tl.cex = 0.7,
                   type = "upper",
                   sig.level = 0.05,
                   insig = "blank",
                   order = "hclust",
                   diag = FALSE)
}
corr_plot <- ggdraw(p1_formula)
saveRDS(corr_plot,file = "plots/gene_corr_plot.RDS")
ggsave(plot = corr_plot,filename = "plots/gene_corr_plot.png",width = 8,height = 8,units = "in",dpi = 300)
corr_plot
```

```{r change_cor_mag_select_genes}
assign.color <- function(dat = DAT, color = col) {
        newcorr <- (dat + 1)/2
        newcorr[newcorr <= 0] <- 0
        newcorr[newcorr >= 1] <- 1 - 1e-16
        color[floor(newcorr * length(color)) + 1]
}

panel.lm =  function (x, y){
    t <- cor.test(x,y,method = "spearman")
    col.set <- ifelse(t$p.value < 0.05,"red","black")
    points(x, y,col = col.set)
    abline(lm(y~x), col="blue")
}

panel.cor <- function(x, y){
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  t <- cor.test(x,y,method = "spearman")
  r <- t$estimate
  r <- round(r,digits = 2)
  col.set <- ifelse(t$p.value < 0.05,assign.color(t$estimate,plot_colours(200)),"white")
  rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = col.set)
  txt <- format(r, digits = 2)[1]
  text(0.5, 0.5, txt, cex = 1.1, font = 4,col = "black")
}

p2_formula <- ~{
  pairs(cor_data,upper.panel = panel.lm,lower.panel = panel.cor,cex.labels = 0.75)
}
pairs_plot <- ggdraw(p2_formula)
saveRDS(pairs_plot,file = "plots/gene_pairs_plot.RDS")

png(filename = "plots/gene_pairs_plot.png",width = 10,height = 8,units = "in",res = 300)
pairs(cor_data,upper.panel = panel.lm,lower.panel = panel.cor,cex.labels = 0.75)
dev.off()
```

```{r change_cor_mat_all_genes}
paired_changes_all <- get_patient_changes(data = copies,
                               genes = protein_coding_genes$hgnc_symbol,
                               patients = unique(meta.data$PATIENT_ID[meta.data$paired == TRUE]))

norm_paired_changes_all <- paired_changes_all %>%
  group_by(PATIENT_ID,group,gene) %>%
  summarise(value=median(value)) %>%
  pivot_wider(values_from = value, names_from=group) %>%
  mutate(change=rlps-arx) %>%
  filter(!PATIENT_ID %in% ploidy_change_samples)
```

```{r normalised_changes_stats_all}
norm_paired_change_summary_all <- norm_paired_changes_all %>%
                        mutate(shift = ifelse(change < 0, "loss", "gain")) %>%
                        group_by(shift) %>%
                        dplyr::select(-gene)

norm_paired_change_summary_median_all <- norm_paired_change_summary_all %>%
                                group_by(PATIENT_ID) %>%
                                summarise(median=median(change),min=min(change),max=max(change)) %>%
                                arrange(median)

# ggplot(norm_paired_change_summary_median_all) +
#   geom_point(aes(PATIENT_ID,median,color=ifelse(median > 0,"gain","loss"))) +
#   geom_linerange(aes(x = PATIENT_ID,ymin=min,ymax=max,color=ifelse(median > 0,"gain","loss"))) +
#   coord_flip()

norm_paired_change_summary_all$PATIENT_ID <- factor(norm_paired_change_summary_all$PATIENT_ID,levels = unique(norm_paired_change_summary_median_all$PATIENT_ID))

ggplot(norm_paired_change_summary_all) +
  geom_point(aes(change,PATIENT_ID,color=shift),alpha = 0.3) +
  geom_boxplot(aes(change,PATIENT_ID),outlier.alpha = 0) +
  theme_bw()
```

```{r norm_cor_all}
cor_data_all <- norm_paired_changes_all %>%
  dplyr::select(-c(arx,rlps)) %>%
  pivot_wider(id_cols = PATIENT_ID,names_from = gene,values_from = change) %>%
  column_to_rownames(var = "PATIENT_ID")
```

```{r change_cor_mat_all_genes_sub}
cor_data_all_vars <- apply(cor_data_all,MARGIN = 2,sd)
topvar <- quantile(cor_data_all_vars,probs = 0.80)
cor_data_all_vars_top <- names(cor_data_all_vars[cor_data_all_vars >= topvar])
cor_data_all_sub <- cor_data_all[names(cor_data_all) %in% cor_data_all_vars_top]
```

```{r change_cor_heatmap_all_genes_sub}
chr_colors_sub <- ifelse(protein_coding_genes$chromosome_name[protein_coding_genes$hgnc_symbol %in% colnames(cor_data_all_sub)] %% 2 == 0,"grey75","grey25")
gene_change_ordering_sub <- match(protein_coding_genes$hgnc_symbol[protein_coding_genes$hgnc_symbol %in% colnames(cor_data_all_sub)],colnames(cor_data_all_sub))

ann_age <- patient_data$age[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]
ann_prior <- patient_data$pre_reg_chemo[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]
ann_pt <- patient_data$pt_sensitivity_at_reg[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]
ann_stage <- patient_data$tumour_stage_at_diagnosis[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]

col_ann <- HeatmapAnnotation(age=ann_age,
                             prior_lines=as.factor(ann_prior),
                             platinum_status=ann_pt,
                             stage=as.factor(ann_stage),
                             col = list(age = circlize::colorRamp2(c(1,100),colors = c("white","purple")),
                                        prior_lines = c("1"="#e5f5f9","2"="#99d8c9","4"="#2ca25f"),
                                        platinum_status = colour_palettes$resist_sensitive,
                                        stage = c("1"="#fef0d9","2"="#fdcc8a","3"="#fc8d59","4"="#d7301f")),
                             gp = gpar(col = "black"))
#row_ann <- rowAnnotation(chromosome=chr_colors_sub)

pdf(file = "plots/gene_change_heatmap.pdf",width = 10,height = 8)
hmPlot <- ComplexHeatmap::Heatmap(t(cor_data_all_sub),
                        name = "Copy number change",
                        top_annotation = col_ann,
                        #right_annotation = row_ann,
                        row_order = gene_change_ordering_sub, 
                        show_row_dend = FALSE,
                        show_row_names = FALSE,
                        use_raster = TRUE)

draw(hmPlot)
dev.off()
```

```{r umap}
umap.data <- umap(cor_data_all_sub)

ggplot(as.data.frame(umap.data$layout)) +
  geom_point(aes(umap.data$layout[,1],umap.data$layout[,2])) +
  xlab("dim 1") +
  ylab("dim 2") +
  labs(title = "Gene change UMAP") +
  theme_bw()
```

```{r heatmap_all_genes_sub_noPloidyChange}
kclust.est <- data.frame("clusters"=as.numeric(),"tot.withinss"=as.numeric())
for(i in 1:10){
  n <- c()
  n <- append(n,kmeans(cor_data_all_sub,centers = i,iter.max = 20)$tot.withinss)
  kclust.est <- rbind(kclust.est,data.frame("clusters"=i,"tot.withinss"=n))
}

ggplot(kclust.est) +
  geom_point(aes(clusters,tot.withinss)) +
  geom_line(aes(clusters,tot.withinss)) +
  theme_bw()

kmean.clust <- kmeans(cor_data_all_sub,centers = 6)

ggplot(as.data.frame(umap.data$layout)) +
  geom_point(aes(umap.data$layout[,1],umap.data$layout[,2],color = as.factor(kmean.clust$cluster))) +
  xlab("dim 1") +
  ylab("dim 2") +
  labs(title = "Gene change UMAP") +
  theme_bw()
```

```{r patient_vignettes}

```

- Plot gene per chromosome from most variable subset and GOterm analysis
- Pattern of patient-level alterations 
- Correlation / co-occurrence with SNV data

### Session info

```{r sessioninfo}
sessionInfo()
```

