---
title: "Analysis of absolute CNAs - BriTROC 30kb - focal (patient specific)"
author: "Philip Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: TRUE
    toc_float: TRUE
---

## Initial steps

### knitr options

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries

```{r load_libraries}
# Load libraries
suppressPackageStartupMessages(suppressWarnings(library(Biobase)))
suppressPackageStartupMessages(suppressWarnings(library(QDNAseqmod)))
suppressPackageStartupMessages(suppressWarnings(library(tidyverse)))
suppressPackageStartupMessages(suppressWarnings(library(umap)))
suppressPackageStartupMessages(suppressWarnings(library(cowplot)))
suppressPackageStartupMessages(suppressWarnings(library(kableExtra)))
suppressPackageStartupMessages(suppressWarnings(library(biomaRt)))
suppressPackageStartupMessages(suppressWarnings(library(ComplexHeatmap)))
suppressPackageStartupMessages(suppressWarnings(library(ggsignif)))
suppressPackageStartupMessages(suppressWarnings(library(ggrepel)))
suppressPackageStartupMessages(suppressWarnings(library(RColorBrewer)))
suppressPackageStartupMessages(suppressWarnings(library(gridExtra)))
```

### Assign int. functions

```{r functions}
source("data/focal_cna_analysis_funcs.R")
source("../../colour_palettes.R")
```

## Load data

### Read meta data

Meta data for all samples with fitted absolute copy number profiles, as well as genomic loci/annotation data for 16 frequently altered genes linked to high grade serous ovarian cancer.

```{r read_meta}
meta.data <- read.table("../../copy_number_signatures/britroc_30kb_signature_data_meta.tsv",header = TRUE,sep = "\t")
patient_data <- read.table("../../britroc_cohort_patient_data.tsv",header = T,sep = "\t")

cytoband <- read.table("data/hg19_cytobands.tsv",header = T,sep = "\t")
cytoband$length <- cytoband$end - cytoband$start

ploidy_change_samples <- read.table(file = "../broad_analysis/data/ploidy_change_pats.tsv",header = T,sep = "\t")
ploidy_change_samples <- unique(ploidy_change_samples$PATIENT_ID)

#signatures
sig.exposures <- read.table("../../copy_number_signatures/britroc_30kb_signature_data_sig_exposures.tsv",header = TRUE,sep = "\t")
colnames(sig.exposures) <- gsub(colnames(sig.exposures),pattern = "\\.",replacement = "-")
signature.table <- sig.exposures %>%
                     #rownames_to_column(var = "signature") %>%
                     pivot_longer(cols = 2:ncol(.),names_to = "SAMPLE_ID",values_to = "exposure") %>%
                     mutate(PATIENT_ID = meta.data$PATIENT_ID[match(SAMPLE_ID,meta.data$SAMPLE_ID)]) %>%
                     mutate(group = meta.data$group[match(SAMPLE_ID,meta.data$SAMPLE_ID)]) %>%
                     mutate(group = case_when(group == "arx" ~ "diagnosis",group == "rlps" ~ "relapse")) %>%
                     rename("signature"="rowname")
#ith
ith <- read.table("../heterogeneity/data/ith_results_table.tsv",header = TRUE,sep = "\t")

# BRCA
```

Frequently altered genes are as listed below:

```{r load_genes}
focal_genes_table <- read.table("data/focal_CNA_pseudo_bed.txt",header = T,sep = "\t")
tcga_gene_rates <- read.table("data/TCGA_gene_CNA_rates.txt",header = T,sep = "\t")
tcga_gene_rates <- tcga_gene_rates[,-2]
tcga_gene_rates <- tcga_gene_rates[,c(1,3,2)]
colnames(tcga_gene_rates) <- c("gene","rate","group")
tcga_gene_rates$group <- gsub(pattern = "HOMDEL",replacement = "DEL",x = tcga_gene_rates$group)
```

```{r freq_gene_list}
kable(focal_genes_table) %>%
  kable_styling() %>%
  scroll_box(height = "200px")
```

### Read absolute data

Absolute copy number profiles are loaded in QDNAseq object format and unused/dropped samples are removed. 

```{r read_abs}
abs_data <- readRDS("../../absolute_POST_down_sampling/britroc_30kb_ds_absCopyNumber.rds")
abs_data <- abs_data[,which(colnames(abs_data) %in% meta.data$SAMPLE_ID[meta.data$use == "TRUE"])]
```

Ploidy as defined by the grid search process for each sample are added to a named vector for use in downstream filtering of copy number states.

```{r set_ploidy}
ploidys <- pData(abs_data)$ploidy
names(ploidys) <- pData(abs_data)$name
```

Using ensembl biomart API, a full list of genes are downloaded and subset to those classified as protein-coding with primary/canonical transcripts, then further limited to autosomes as sex chromosomes are not included in the CNA profile data.

```{r biomart}
if(file.exists("data/protein_coding_genes.tsv")){
	protein_coding_genes <- read.table("data/protein_coding_genes.tsv",header=T,sep="\t")
} else {
	ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl",GRCh = 37)
	protein_coding_genes <- getBM(attributes = c("chromosome_name","start_position","end_position","hgnc_symbol"),
                                             filters = "biotype",
                                             values = "protein_coding",
                                             mart = ensembl,
                                             useCache = FALSE)
	protein_coding_genes <- protein_coding_genes[protein_coding_genes$chromosome_name %in% seq.int(1,22,1),]
	protein_coding_genes <- protein_coding_genes[protein_coding_genes$hgnc_symbol != "",]
	write.table(protein_coding_genes,"data/protein_coding_genes.tsv",col.names=T,row.names=F,sep="\t")
}
protein_coding_genes$chromosome_name <- as.numeric(protein_coding_genes$chromosome_name)
protein_coding_genes <- protein_coding_genes[order(protein_coding_genes$chromosome_name,protein_coding_genes$start_position),]
```

#### Extract gene-level CNA

Here quick loading of pre-extracted gene-level CNA values are loaded as this function can take a large amount of time to generate. In an instance where this data is not available a function is executed to tabulate the segment value for each gene across each sample in the BriTROC study.

```{r abs_only_focal}
if(file.exists("data/extracted_output.RDS")){
  extracted <- readRDS("data/extracted_output.RDS")
} else {
  extracted <- extract_abs_segs(abs_data = abs_data,genes = protein_coding_genes) # long execution times
  saveRDS(extracted,file = "data/extracted_output.RDS") 
}
```

```{r copies_ext}
copies <- extracted
```

```{r patient_level_changes,fig.width=12}
paired_changes <- get_patient_changes(data = copies,
                                              genes = focal_genes_table$Gene.name,
                                              patients = unique(meta.data$PATIENT_ID[meta.data$paired == TRUE]))

norm_paired_changes <- paired_changes %>%
  group_by(PATIENT_ID,group,gene) %>%
  summarise(value=median(value)) %>%
  pivot_wider(values_from = value, names_from=group) %>%
  mutate(change=rlps-arx)
```

### Patient-specific changes

```{r pat_specific_justification}
pat_ordering <- norm_paired_changes %>%
                  group_by(PATIENT_ID) %>%
                  summarise(across(change,sum)) %>%
                  arrange(change)

norm_paired_changes$PATIENT_ID <- factor(norm_paired_changes$PATIENT_ID,
                                         levels = as.character(rev(pat_ordering$PATIENT_ID)))

norm_paired_changes_plot <- ggplot(norm_paired_changes) +
  geom_hline(yintercept = 0) +
  geom_col(aes(x = PATIENT_ID,y = change,group = gene,fill=ifelse(PATIENT_ID %in% ploidy_change_samples,"changed","unchanged"))) +
  facet_wrap(. ~ gene,scales = "free_y",ncol = 3) +
  ylab("copy number change") +
  xlab("patient") +
  scale_fill_manual(name = "ploidy change",values = c("firebrick2","grey50")) +
  theme_bw() +
  theme(axis.text.x = element_blank(),axis.ticks = element_blank())

ggsave(plot = norm_paired_changes_plot,filename = "plots/norm_paired_changes_plot.png",width = 8,height = 8,units = "in",dpi = 300)
ggsave(plot = norm_paired_changes_plot,filename = "plots/norm_paired_changes_plot.pdf",width = 8,height = 8,units = "in",dpi = 300)
norm_paired_changes_plot
```

```{r source_data_SM3}
write.table(x = norm_paired_changes,
            file = "../../source_data/figure_SM3.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r normalised_changes_stats_genelist}
norm_paired_change_summary <- norm_paired_changes %>%
                        mutate(shift = ifelse(change < 0, "loss", "gain")) %>%
                        group_by(shift) %>%
                        dplyr::select(-gene)

norm_change_summary_median <- norm_paired_change_summary %>%
                                group_by(PATIENT_ID) %>%
                                summarise(median=median(change)) %>%
                                arrange(median)

norm_paired_change_summary$PATIENT_ID <- factor(norm_paired_change_summary$PATIENT_ID,levels = unique(norm_change_summary_median$PATIENT_ID))

ggplot(norm_paired_change_summary) +
  geom_point(aes(change,PATIENT_ID,color=shift),alpha = 0.5) +
  geom_boxplot(aes(change,PATIENT_ID,fill=shift),outlier.alpha = 0) +
  theme_bw()
```

```{r change_cor_mat_select_genes}
cor_data <- norm_paired_changes %>%
  filter(!PATIENT_ID %in% ploidy_change_samples) %>%
  dplyr::select(-c(arx,rlps)) %>%
  pivot_wider(id_cols = PATIENT_ID,names_from = gene,values_from = change) %>%
  column_to_rownames(var = "PATIENT_ID")

cor_matrix <- Hmisc::rcorr(x = as.matrix(cor_data),type = "spearman")

plot_colours <- colorRampPalette(rev(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA")))
corrplot::corrplot(cor_matrix$r,col = col(matrix(200)),mar = c(1, 0, 1, 0),
                   method = "color",
                   p.mat = cor_matrix$P,
                   addCoef.col = "black",
                   number.cex = 0.7,
                   tl.col = "black",
                   tl.srt = 90,
                   tl.cex = 0.7,
                   type = "upper",
                   sig.level = 0.01,
                   insig = "blank",order = "hclust",
                   diag = FALSE)

p1_formula <- ~{
  corrplot::corrplot(corr = cor_matrix$r,number.digits = 2,col = plot_colours(200),
                   mar = c(1, 0, 1, 0),
                   method = "color",addgrid.col = "grey90",
                   p.mat = cor_matrix$P,
                   addCoef.col = "black",
                   number.cex = 0.6,
                   tl.col = "black",
                   tl.srt = 90,
                   tl.cex = 0.7,
                   type = "upper",
                   sig.level = 0.05,
                   insig = "blank",
                   order = "hclust",
                   diag = FALSE)
}
corr_plot <- ggdraw(p1_formula)
saveRDS(corr_plot,file = "plots/gene_corr_plot.RDS")
ggsave(plot = corr_plot,filename = "plots/gene_corr_plot.png",width = 8,height = 8,units = "in",dpi = 300)
ggsave(plot = corr_plot,filename = "plots/gene_corr_plot.pdf",width = 8,height = 8,units = "in",dpi = 300)
corr_plot
```

```{r change_cor_mag_select_genes}
assign.color <- function(dat = DAT, color = col) {
        newcorr <- (dat + 1)/2
        newcorr[newcorr <= 0] <- 0
        newcorr[newcorr >= 1] <- 1 - 1e-16
        color[floor(newcorr * length(color)) + 1]
}

panel.lm =  function (x, y){
    t <- cor.test(x,y,method = "spearman")
    col.set <- ifelse(t$p.value < 0.05,"red","black")
    points(x, y,col = col.set)
    abline(lm(y~x), col="blue")
}

panel.cor <- function(x, y){
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  t <- cor.test(x,y,method = "spearman")
  r <- t$estimate
  r <- round(r,digits = 2)
  col.set <- ifelse(t$p.value < 0.05,assign.color(t$estimate,plot_colours(200)),"white")
  rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = col.set)
  txt <- format(r, digits = 2)[1]
  text(0.5, 0.5, txt, cex = 1.1, font = 4,col = "black")
}

p2_formula <- ~{
  pairs(cor_data,upper.panel = panel.lm,lower.panel = panel.cor,cex.labels = 0.75)
}
pairs_plot <- ggdraw(p2_formula)
saveRDS(pairs_plot,file = "plots/gene_pairs_plot.RDS")

png(filename = "plots/gene_pairs_plot.png",width = 10,height = 8,units = "in",res = 300)
pairs(cor_data,upper.panel = panel.lm,lower.panel = panel.cor,cex.labels = 0.75)
dev.off()

pdf(file = "plots/gene_pairs_plot.pdf",width = 10,height = 8)
pairs(cor_data,upper.panel = panel.lm,lower.panel = panel.cor,cex.labels = 0.75)
dev.off()

pairs(cor_data,upper.panel = panel.lm,lower.panel = panel.cor,cex.labels = 0.75)
```

```{r source_data_S21}
write.table(x = cor_data,
            file = "../../source_data/figure_S21.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r change_cor_mat_all_genes}
paired_changes_all <- get_patient_changes(data = copies,
                               genes = protein_coding_genes$hgnc_symbol,
                               patients = unique(meta.data$PATIENT_ID[meta.data$paired == TRUE]))

norm_paired_changes_all <- paired_changes_all %>%
  group_by(PATIENT_ID,group,gene) %>%
  summarise(value=median(value)) %>%
  pivot_wider(values_from = value, names_from=group) %>%
  mutate(change=rlps-arx) %>%
  filter(!PATIENT_ID %in% ploidy_change_samples)
```

```{r normalised_changes_stats_all}
norm_paired_change_summary_all <- norm_paired_changes_all %>%
                        mutate(shift = ifelse(change < 0, "loss", "gain")) %>%
                        group_by(shift) %>%
                        dplyr::select(-gene)

norm_paired_change_summary_median_all <- norm_paired_change_summary_all %>%
                                group_by(PATIENT_ID) %>%
                                summarise(median=median(change),min=min(change),max=max(change)) %>%
                                arrange(median)

# ggplot(norm_paired_change_summary_median_all) +
#   geom_point(aes(PATIENT_ID,median,color=ifelse(median > 0,"gain","loss"))) +
#   geom_linerange(aes(x = PATIENT_ID,ymin=min,ymax=max,color=ifelse(median > 0,"gain","loss"))) +
#   coord_flip()

norm_paired_change_summary_all$PATIENT_ID <- factor(norm_paired_change_summary_all$PATIENT_ID,levels = unique(norm_paired_change_summary_median_all$PATIENT_ID))

ggplot(norm_paired_change_summary_all) +
  geom_point(aes(change,PATIENT_ID,color=shift),alpha = 0.3) +
  geom_boxplot(aes(change,PATIENT_ID),outlier.alpha = 0) +
  theme_bw()
```

```{r ploidy_norm_change}
norm_paired_changes_ploidy <- norm_paired_changes_all %>%
  mutate(ploidy = meta.data$ploidy[match(PATIENT_ID,meta.data$PATIENT_ID)]) %>%
  filter(gene %in% focal_genes_table$Gene.name) %>%
  mutate(arx_norm = arx - ploidy) %>%
  mutate(rlps_norm = rlps - ploidy) %>%
  mutate(gene = factor(gene,levels = unique(focal_genes_table$Gene.name))) %>%
  mutate(class = focal_genes_table$expected[match(gene,focal_genes_table$Gene.name)]) %>%
  dplyr::select(-arx,-rlps) %>%
  pivot_longer(cols = 5:6,names_to = "group",values_to = "copy number") %>%
  mutate(group = case_when(group == "arx_norm" ~ "diagnosis",group == "rlps_norm" ~ "relapse"))

ploidy_norm_gene_cn_plot <- ggplot(norm_paired_changes_ploidy) +
  geom_hline(yintercept = 0) +
  geom_line(aes(group,`copy number`,group=PATIENT_ID),alpha = 0.15) +
  geom_jitter(aes(group,`copy number`,color=group),alpha = 0.3,
              position = position_dodge(width = 0.6)) +
  geom_violin(aes(group,`copy number`,color=group),fill=NA) +
  ylab("copy number (ploidy normalised)") +
  facet_wrap(. ~ gene,scales = "free") +
  scale_color_manual(values = colour_palettes$diagnosis_relapse) +
  theme_bw()

saveRDS(ploidy_norm_gene_cn_plot,file = "plots/ploidy_norm_gene_cn_plot.RDS")
ggsave(plot = ploidy_norm_gene_cn_plot,filename = "plots/ploidy_norm_gene_cn_plot.png",width = 8,height = 6,units = "in",dpi = 300)
ggsave(plot = ploidy_norm_gene_cn_plot,filename = "plots/ploidy_norm_gene_cn_plot.pdf",width = 8,height = 6,units = "in",dpi = 300)
ploidy_norm_gene_cn_plot
```

```{r source_data_S23}
write.table(x = norm_paired_changes_ploidy,
            file = "../../source_data/figure_S23.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r norm_cor_all}
cor_data_all <- norm_paired_changes_all %>%
  dplyr::select(-c(arx,rlps)) %>%
  pivot_wider(id_cols = PATIENT_ID,names_from = gene,values_from = change) %>%
  column_to_rownames(var = "PATIENT_ID")
```

```{r change_cor_mat_all_genes_sub}
cor_data_all_vars <- apply(cor_data_all,MARGIN = 2,sd)
topvar <- quantile(cor_data_all_vars,probs = 0.80)
cor_data_all_vars_top <- names(cor_data_all_vars[cor_data_all_vars >= topvar])
cor_data_all_sub <- cor_data_all[names(cor_data_all) %in% cor_data_all_vars_top]
```

```{r change_cor_heatmap_all_genes_sub}
chr_colors_sub <- ifelse(protein_coding_genes$chromosome_name[protein_coding_genes$hgnc_symbol %in% colnames(cor_data_all_sub)] %% 2 == 0,"grey75","grey25")
gene_change_ordering_sub <- match(protein_coding_genes$hgnc_symbol[protein_coding_genes$hgnc_symbol %in% colnames(cor_data_all_sub)],colnames(cor_data_all_sub))

ann_age <- patient_data$age[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]
ann_prior <- patient_data$pre_reg_chemo[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]
ann_pt <- patient_data$pt_sensitivity_at_reg[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]
ann_stage <- patient_data$tumour_stage_at_diagnosis[match(rownames(cor_data_all_sub),
                                             paste0("BRITROC-",patient_data$britroc_number))]


col_ann <- HeatmapAnnotation(age=ann_age,
                             prior_lines=as.factor(ann_prior),
                             platinum_status=ann_pt,
                             stage=as.factor(ann_stage),
                             col = list(age = circlize::colorRamp2(c(1,100),colors = c("white","purple")),
                                        prior_lines = colour_palettes$prior_lines,
                                        platinum_status = colour_palettes$resist_sensitive,
                                        stage = colour_palettes$stage),
                             gp = gpar(col = "black"))
#row_ann <- rowAnnotation(chromosome=chr_colors_sub)

hmPlot <- ComplexHeatmap::Heatmap(t(cor_data_all_sub),
                        name = "Copy number change",
                        top_annotation = col_ann,
                        show_column_names = FALSE,
                        #right_annotation = row_ann,
                        row_order = gene_change_ordering_sub, 
                        show_row_dend = FALSE,
                        show_row_names = FALSE,
                        use_raster = TRUE)

pdf(file = "plots/gene_change_heatmap.pdf",width = 10,height = 8)
draw(hmPlot)
dev.off()

png(file = "plots/gene_change_heatmap.png",width = 10,height = 8,units = "in",res = 300)
draw(hmPlot)
dev.off()

heatmapGG <- ggdraw(grid.grabExpr(draw(hmPlot)))
saveRDS(heatmapGG,file = "plots/gene_change_heatmap.RDS")
hmPlot
```

```{r source_data_S20}
write.table(x = cor_data_all_sub,
            file = "../../source_data/figure_S20.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```


```{r gene_change_umap_and_kclusts}
umap.data <- umap(cor_data_all_sub)

u1 <- ggplot(as.data.frame(umap.data$layout)) +
  geom_point(aes(umap.data$layout[,1],umap.data$layout[,2])) +
  xlab("dim 1") +
  ylab("dim 2") +
  labs(title = "UMAP") +
  theme_bw()

kclust.est <- data.frame("clusters"=as.numeric(),"tot.withinss"=as.numeric())
for(i in 1:10){
  n <- c()
  n <- append(n,kmeans(cor_data_all_sub,centers = i,iter.max = 20)$tot.withinss)
  kclust.est <- rbind(kclust.est,data.frame("clusters"=i,"tot.withinss"=n))
}

u2 <- ggplot(kclust.est) +
  geom_point(aes(clusters,tot.withinss)) +
  geom_line(aes(clusters,tot.withinss)) +
  xlab("cluster") +
  ylab("total within sum of squares") +
  labs(title = "K optimisation") +
  theme_bw()

kmean.clust <- kmeans(cor_data_all_sub,centers = 6)

u3 <- ggplot(as.data.frame(umap.data$layout)) +
  geom_point(aes(umap.data$layout[,1],umap.data$layout[,2],
                 color = as.factor(kmean.clust$cluster))) +
  xlab("dim 1") +
  ylab("dim 2") +
  labs(title = "clustered UMAP") +
  scale_color_discrete(name = "cluster") +
  theme_bw() +
  theme(legend.position = "none")

umap_legend <- get_legend(
  u3 + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)
umap_plot <- plot_grid(u1,plot_grid(u2,u3,ncol = 2,labels = c("B","C")),umap_legend,nrow = 3,labels = c("A","",""),rel_heights = c(0.45,0.45,0.1))

saveRDS(umap_plot,file = "plots/umap_plot.RDS")
ggsave(plot = umap_plot,filename = "plots/umap_plot.png",width = 6,height = 6,units = "in",dpi = 300)
ggsave(plot = umap_plot,filename = "plots/umap_plot.pdf",width = 6,height = 6,units = "in",dpi = 300)
umap_plot
```

```{r source_data_S22A}
write.table(x = as.data.frame(umap.data$layout),
            file = "../../source_data/figure_S22A.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r source_data_S22B}
write.table(x = kclust.est,
            file = "../../source_data/figure_S22B.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r source_data_S22C}
write.table(x = as.data.frame(umap.data$layout),
            file = "../../source_data/figure_S22C.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r recalc_paired_diffs}
paired_diffs <- get_paired_genome_diffs(data = abs_data[,colnames(abs_data) %in% meta.data$SAMPLE_ID[meta.data$paired == "TRUE"]],
                                        method = "median",
                                        arx_samples = meta.data$SAMPLE_ID[meta.data$paired == T & meta.data$group == "arx"],
                                        rlps_samples = meta.data$SAMPLE_ID[meta.data$paired == T & meta.data$group == "rlps"]
                                        )
paired_diffs[paired_diffs <= 0] <- 0.0005
diff_matrix <- as.matrix(paired_diffs[,c(((ncol(paired_diffs)/2)+1):ncol(paired_diffs))]) - as.matrix(paired_diffs[,1:(ncol(paired_diffs)/2)])
colnames(diff_matrix) <- gsub(colnames(diff_matrix),pattern = "_rlps",replacement = "")
```

```{r norm_paired_changes_all}
norm_paired_changes_all <- norm_paired_changes_all %>%
                            mutate(ploidy_changed = ifelse(PATIENT_ID %in% ploidy_change_samples,"changed","unchanged")) %>%
                            arrange(change) %>%
                            filter(ploidy_changed == "unchanged") %>%
                            mutate(chr = protein_coding_genes$chromosome_name[match(gene,protein_coding_genes$hgnc_symbol)])
norm_paired_changes_all
norm_paired_changes_sub <- norm_paired_changes_all[norm_paired_changes_all$gene %in% focal_genes_table$Gene.name,]
norm_paired_changes_sub[order(norm_paired_changes_sub$change),]
```

```{r gene_change_spread}
norm_paired_changes_all$status <- patient_data$pt_sensitivity_at_reg[match(norm_paired_changes_all$PATIENT_ID,
                                                                           paste0("BRITROC-",patient_data$britroc_number))]

norm_paired_changes_all$prior_lines <- patient_data$pre_reg_chemo[match(norm_paired_changes_all$PATIENT_ID,
                                                           paste0("BRITROC-",patient_data$britroc_number))]

#norm_paired_changes_all <- norm_paired_changes_all[sample(c(1:nrow(norm_paired_changes_all)),size = 10000,replace = F),]

pt_jitter <- ggplot(norm_paired_changes_all) +
  geom_jitter(aes(status,change,group=status,color=status),
              alpha=0.2,
              position = position_dodge2(width = 0.3)) +
  geom_signif(aes(status,change),comparisons = list(c("resistant","sensitive"))) +
  scale_color_manual(values = colour_palettes$resist_sensitive) +
  theme_bw() +
  theme(legend.position = "none")

pt_point <- ggplot(norm_paired_changes_all) +
  geom_point(aes(arx,rlps,color=status),alpha=0.2) +
  geom_text_repel(data = norm_paired_changes_all[abs(norm_paired_changes_all$change) > 30,],
                            aes(arx,rlps,color=status,label=gene),
                  box.padding = 0.5, max.overlaps = Inf) +
  ylab("relapse") +
  xlab("diagnosis") +
  theme_bw() +
  theme(legend.position = "bottom")

pt_spread_plot <- plot_grid(pt_jitter,pt_point,nrow = 2,labels = c("A","B"))
ggsave2(pt_spread_plot,filename = "plots/pt_spread_plot.pdf",height = 8,width = 7)
ggsave2(pt_spread_plot,filename = "plots/pt_spread_plot.png",height = 8,width = 7,dpi = 300)
pt_spread_plot
```

```{r source_data_S25}
write.table(x = norm_paired_changes_all,
            file = "../../source_data/figure_S25.tsv",
            quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r gene_change_spread_pat_strat}
ggplot(norm_paired_changes_all) +
  geom_jitter(aes(status,change,group=status,color=status),
              alpha=0.2,
              position = position_dodge2(width = 0.3)) +
  geom_signif(aes(status,change),comparisons = list(c("resistant","sensitive"))) +
  facet_wrap(. ~ PATIENT_ID) +
  scale_color_manual(values = colour_palettes$resist_sensitive) +
  theme_bw() +
  theme(legend.position = "none")

ggplot(norm_paired_changes_all) +
  geom_point(aes(arx,rlps,color=status),alpha=0.2) +
  #geom_text_repel(data = norm_paired_changes_all[abs(norm_paired_changes_all$change) > 30,],
    #                        aes(arx,rlps,color=status,label=gene),
     #             box.padding = 0.5, max.overlaps = Inf) +
  facet_wrap(. ~ PATIENT_ID) +
  ylab("relapse") +
  xlab("diagnosis") +
  theme_bw() +
  theme(legend.position = "bottom")
```
 
#### Patient vignettes

Patient-specific alterations, while difficult to associate with specific clinical events or features as a singular observation, did demonstrate biologically interesting genomic alterations. For example,  BRITROC-65 & BRITROC-37 demonstrated extreme gains of KRAS.Whereas other patients had notable losses of clinically relevant or frequently altered genes such as BRITROC-36 which had a 16 copy loss of AKT2 and 6 copy loss of NF1, and BRITROC-242 & BRITROC-246 which demonstrated more than 10 copy loss of CCNE1.

```{r plot_vigs_example}
pat_vig_65 <- get_patient_vig(diff_matrix = diff_matrix,pat.name = "BRITROC-65")
pat_vig_65
```

Plot all patient vignettes

```{r plot_vigs,warning=FALSE}
pats <- colnames(diff_matrix)

plot_list <- list()
for(x in pats){
    plt <- list(get_patient_vig(diff_matrix = diff_matrix,pat.name = x))
    plot_list <- append(plot_list,plt)
}

ggsave(filename = "plots/patient_vignette_plots.pdf",plot = marrangeGrob(plot_list, nrow=1, ncol=1),width = 8, height = 11.3136)
```

### Session info

```{r sessioninfo}
sessionInfo()
```
